--- rbot.orig/lib/rbot/ircsocket.rb	2006-07-29 10:24:49.000000000 -0700
+++ rbot/lib/rbot/ircsocket.rb	2007-01-22 12:52:23.000000000 -0800
@@ -264,6 +264,15 @@
       else
         @sock=TCPSocket.new(@server, @port)
       end
+
+      require 'openssl'
+      ssl_context = OpenSSL::SSL::SSLContext.new()
+      ssl_context.verify_mode = OpenSSL::SSL::VERIFY_NONE
+      @rawsock = @sock
+      @sock = OpenSSL::SSL::SSLSocket.new(@rawsock, ssl_context)
+      @sock.sync_close = true
+      @sock.connect
+
       @qthread = false
       @qmutex = Mutex.new
       @sendq = MessageQueue.new
@@ -430,7 +439,14 @@
 
     # shutdown the connection to the server
     def shutdown(how=2)
-      @sock.shutdown(how) unless @sock.nil?
+      return unless connected?
+      begin
+        @sock.close
+      rescue => err
+        error "error while shutting down: #{err.inspect}"
+        debug err.backtrace.join("\n")
+      end
+      @rawsock = nil
       @sock = nil
       @burst = 0
     end
@@ -445,7 +461,7 @@
         if @sock.nil?
           error "SEND attempted on closed socket"
         else
-          @sock.send(message + "\n",0)
+          @sock.puts message
           @last_send = Time.new
           @lines_sent += 1
           @burst += 1
