// vuntangle/src/components/UGrid/props.js
// import translations
import i18n from '../../plugins/vue-i18n'

export default {
  id: { type: String, required: true },
  rowNodeId: { type: String, required: false, default: null },
  rowData: { type: Array, default: () => [] }, // the row data
  columnDefs: { type: Array, default: () => [] }, // the columns definitions specific for each grid
  customGridOptions: { type: Object, required: false, default: null }, // overwrite/extend any grid options
  customDefaultColOptions: { type: Object, required: false, default: null }, // overwrite/extend any default colDef options
  frameworkComponents: { type: Object, default: () => {} }, // custom grid framework components
  selectionType: { type: String, default: 'single' }, // 'single', 'singleAction', 'multiAction'

  // synced props
  selection: { type: Array, default: () => [] }, // the selected rows synced with the parent

  /**
   * Flag to do an auto-resize on the columns.  This is used when resizing a widget so the columns
   * will fit into view.
   */
  resizeColumns: { type: Boolean, default: false },

  toolbar: { type: String, default: 'top' }, // toolbar options, 'top', 'bottom', 'float', 'hidden'

  // the default no data message to show inside grid
  noDataMessage: { type: String, default: i18n.t('no_data_available') },

  // conditional features enabled from toolbar or column menu
  enableRefresh: { type: Boolean, default: true },
  enableExportCsv: { type: Boolean, default: true },

  modules: { type: Array, required: false, default: () => [] },

  // does not set oultined grid wrapper
  noBorder: { type: Boolean, default: false },

  // custom ordering functionality
  customOrdering: { type: Boolean, default: false },
}
