// vuntangle/src/plugins/dates.js
/**
 * This plugin is a collection of methods relevant to date conversion and formatting.
 */
import { captureException } from '@sentry/vue'
import i18n from './vue-i18n'

const dates = {
  /**
   * Formats the given date from the server API as a readable date.
   *
   * if bool return empty string
   * if number use
   * if string of number it will work
   * if there are milliseconds on the time, remove them using split
   *
   * @param {any}  apiDate      date returned from a web service call.  May be timestamp, C# serialized date, etc.
   * @param {bool} displayTime
   *
   * @returns {string}  string returned of converted date/time, includes seconds if flag is passed
   */
  formatDateFromApi(apiDate, displayTime = true) {
    // check if the api date came in as 'N/A'
    if (typeof apiDate === 'string' && apiDate.toLowerCase() === 'n/a') {
      return i18n.t('global.not_applicable')
    }

    // check if an api date can be converted to a timestamp
    let dateValue = this.getTimestampFromApi(apiDate)

    // if the api date could not be convert into a timestamp, try to format a date value
    if (dateValue === null) {
      dateValue = this.formatDateValueFromApi(apiDate)
    }

    // try to convert date value to a date object
    const dateObject = this.getDateObject(dateValue)
    if (!dateObject) {
      return ''
    }

    // return the date object as a local format
    return this.formatLocaleDate(dateObject, displayTime)
  },

  /**
   * Compare two dates by converting to a timestamp.  Used in ag-grid column comparator.
   *
   * @param {any} the first date to compare
   * @param {any} the second date to compare
   *
   * @returns {number}
   */
  compareDates(a, b) {
    return this.getComparatorTimestampFromApi(a) - this.getComparatorTimestampFromApi(b)
  },

  /**
   * Get a timestamp from an api date for comparing dates when sorting.
   *
   * @param {any} apiDate The api date may be relative or absolute like 'now', 'never', 'n/a'
   *
   * @returns {number}
   */
  getComparatorTimestampFromApi(apiDate) {
    // check for valid date
    if (!apiDate) {
      return 0
    }

    // check for relative string dates
    if (typeof apiDate === 'string') {
      // check for relative time of 'now'
      const lowercaseApiDate = apiDate.toLowerCase()
      if (lowercaseApiDate === 'now') {
        return new Date().getTime()
      }

      // check for relative time of 'never'
      if (lowercaseApiDate === 'never' || lowercaseApiDate === 'n/a') {
        return 0
      }
    }

    // check if the api date is already in a timestamp format
    const apiDateTimestamp = this.getTimestampFromApi(apiDate)
    if (apiDateTimestamp !== null) {
      return apiDateTimestamp
    }

    // if the apiDate was not a timestamp, try to format a date value and convert to date object
    const dateValue = this.formatDateValueFromApi(apiDate)
    const dateObject = this.getDateObject(dateValue)
    if (!dateObject) {
      return 0
    }

    // return date timestamp
    return dateObject.getTime()
  },

  /**
   * Get a formatted date value from an api date.  This can then be used as the constructor of a javascript date
   * object.
   *
   * @param {any} apiDate
   *
   * @returns {any}
   */
  formatDateValueFromApi(apiDate) {
    // make sure a string was passed for formatting
    if (typeof apiDate !== 'string') {
      return apiDate
    }

    // e.g. 2019-09-20T12:18:24 remove the T (check for numbers after to not remove T from 'Tue' or 'Thu')
    if (apiDate.match(/T\d/)) {
      return apiDate.replace('T', ' ').split('.')[0]
    }

    // looks like it is already formatted, return with stripping timezone
    if (apiDate.includes('-')) {
      return apiDate.split('.')[0]
    }

    // e.g. 11/30/2018 7:00:00 AM
    if (apiDate.includes('/')) {
      return this.convertSlashDateToLocaleReady(apiDate)
    }

    return apiDate
  },

  /**
   * Get a timestamp from an api date.  There are multiple ways an api date may be formatted for a timestamp.
   *
   * @param {any} apiDate
   *
   * @returns {number|null}
   */
  getTimestampFromApi(apiDate) {
    // make sure a value was passed
    if (!apiDate) {
      return null
    }

    let dateValue

    switch (true) {
      // if already number, use it
      case typeof apiDate === 'number':
        dateValue = apiDate
        break
      // if string as timestamp number, parse
      case typeof apiDate === 'string' && !isNaN(apiDate):
        dateValue = parseInt(apiDate)
        break
      // if object as timestamp number, parse date inside object
      case typeof apiDate === 'object':
        dateValue = parseInt(apiDate.$date)
        break
      // if C# serializer DateTime "/Date(1494957608139)/" as number, parse it
      case apiDate.startsWith('/Date'):
        dateValue = parseInt(apiDate.replace('/Date(', '').replace(')/', ''))
        break
      // if none of the cases above, return early
      default:
        return null
    }

    const dateAsString = String(dateValue)
    // if the timestamp is 13 digits or longer, assume already milliseconds.
    // if not, assume seconds and convert to ms
    return dateAsString.length >= 13 ? dateValue : dateValue * 1000
  },

  /**
   * Returns a date object or null if a valid date object could not be created correctly.
   *
   * @param {any} dateValue possible date value to be passed into a date constructor
   *
   * @returns {Date|null}
   */
  getDateObject(dateValue) {
    // make sure a value was passed
    if (!dateValue) {
      return null
    }

    // return null if the date could not be made into a valid object
    const date = new Date(dateValue)
    if (isNaN(date.getTime())) {
      return null
    }

    return date
  },

  /**
   * Returns a locale formatted date from a UTC timestamp
   * Locale formats are defined in /plugins/vue-i18n.js
   * @param date - UTC timestamp milliseconds
   * @param displayTime - boolean to show or not the time part of date
   */
  formatLocaleDate(dateObject, displayTime = true) {
    try {
      return i18n.d(dateObject, displayTime ? 'long' : 'short')
    } catch (e) {
      captureException(e)
      return ''
    }
  },

  /**
   * Returns a date and time string(yyyy-mm-dd hh:mm:ss) for the specified date
   * @param baseDate - date to format as a string
   * @param dayCount - number of days either before or after baseDate to return formatted date string for
   * @returns {string} - date string formatted as yyyy-mm-dd hh:mm:ss
   */
  getDateTimeString(baseDate, dayCount) {
    const date = new Date(new Date().setDate(baseDate.getDate() + dayCount))
    const dateTimeString =
      date.getUTCFullYear() +
      '-' +
      ('0' + (date.getUTCMonth() + 1)).slice(-2) +
      '-' +
      ('0' + date.getUTCDate()).slice(-2) +
      ' ' +
      ('0' + date.getUTCHours()).slice(-2) +
      ':' +
      ('0' + date.getUTCMinutes()).slice(-2) +
      ':' +
      ('0' + date.getUTCSeconds()).slice(-2)
    return dateTimeString
  },

  /**
   * Returns a date string (yyyy-mm-dd) for a date.
   * @param baseDate - date to format as a string
   * @param dayCount - number of days either before or after baseDate to return formatted date string for
   * @returns {string} - date string formatted as yyyy-mm-dd
   */
  getDateString(baseDate, dayCount) {
    const date = new Date(new Date().setDate(baseDate.getDate() + dayCount))
    const dateString =
      date.getUTCFullYear() +
      '-' +
      ('0' + (date.getUTCMonth() + 1)).slice(-2) +
      '-' +
      ('0' + date.getUTCDate()).slice(-2)
    return dateString
  },

  /**
   * Converts a date format of MM/DD/YYYY HH:MM:SS AM/PM to YYYY-MM-DD HH:MM:SS
   * so formatLocaleDate can process properly
   * @param {*} slashdate - date that needs to be converted to proper format
   */
  convertSlashDateToLocaleReady(slashdate) {
    if (!slashdate) {
      return false
    }
    const slashdateSplit = slashdate.split(' ')
    const slashDate = slashdateSplit[0]
    const slashTime = slashdateSplit[1]
    const slashIsAM = slashdateSplit[2] === 'AM'

    // convert date to YYYY-mm-dd
    const dateItems = slashDate.split('/')
    const date = dateItems[2] + '-' + this.addZeroToDate(dateItems[0]) + '-' + this.addZeroToDate(dateItems[1]) + ' '

    // convert time to military time
    let time = slashTime
    if (!slashIsAM) {
      const timeItems = slashTime.split(':')
      const hour = parseInt(timeItems[0]) + 12
      time = hour + ':' + timeItems[1] + ':' + timeItems[2]
    }

    return date + time
  },

  /**
   * Used by convertSlashDateToLocaleReady. A month or day needs at least two digits to be formatted properly
   * this ensures that the month/day has that
   * @param {*} dateItem - month or day that needs to be checked for validity
   */
  addZeroToDate(dateItem) {
    if (dateItem.length === 1) {
      return '0' + dateItem
    }
    return dateItem
  },

  timeZones: [
    { text: '(~UTC-12:00) Etc/GMT+12', value: 'Etc/GMT+12' },
    { text: '(~UTC-11:00) Etc/GMT+11', value: 'Etc/GMT+11' },
    { text: '(~UTC-11:00) Pacific/Midway', value: 'Pacific/Midway', openwrt: 'SST11' },
    { text: '(~UTC-11:00) Pacific/Niue', value: 'Pacific/Niue', openwrt: 'NUT11' },
    { text: '(~UTC-11:00) Pacific/Pago_Pago', value: 'Pacific/Pago_Pago', openwrt: 'SST11' },
    { text: '(~UTC-11:00) Pacific/Samoa', value: 'Pacific/Samoa' },
    { text: '(~UTC-11:00) US/Samoa', value: 'US/Samoa' },
    { text: '(~UTC-10:00) Etc/GMT+10', value: 'Etc/GMT+10' },
    { text: '(~UTC-10:00) Pacific/Honolulu', value: 'Pacific/Honolulu', openwrt: 'HST10' },
    { text: '(~UTC-10:00) Pacific/Johnston', value: 'Pacific/Johnston', openwrt: 'HST10' },
    { text: '(~UTC-10:00) Pacific/Rarotonga', value: 'Pacific/Rarotonga', openwrt: 'CKT10' },
    { text: '(~UTC-10:00) Pacific/Tahiti', value: 'Pacific/Tahiti', openwrt: 'TAHT10' },
    { text: '(~UTC-10:00) SystemV/HST10', value: 'SystemV/HST10' },
    { text: '(~UTC-10:00) US/Hawaii', value: 'US/Hawaii' },
    { text: '(~UTC-10:00) HST', value: 'HST' },
    { text: '(~UTC-09:30) Pacific/Marquesas', value: 'Pacific/Marquesas', openwrt: 'MART9:30' },
    { text: '(~UTC-09:00) America/Adak', value: 'America/Adak', openwrt: 'HAST10HADT,M3.2.0,M11.1.0' },
    { text: '(~UTC-09:00) America/Atka', value: 'America/Atka' },
    { text: '(~UTC-09:00) Etc/GMT+9', value: 'Etc/GMT+9' },
    { text: '(~UTC-09:00) Pacific/Gambier', value: 'Pacific/Gambier', openwrt: 'GAMT9' },
    { text: '(~UTC-09:00) SystemV/YST9', value: 'SystemV/YST9' },
    { text: '(~UTC-09:00) SystemV/YST9YDT', value: 'SystemV/YST9YDT' },
    { text: '(~UTC-09:00) US/Aleutian', value: 'US/Aleutian' },
    { text: '(~UTC-08:00) America/Anchorage', value: 'America/Anchorage', openwrt: 'AKST9AKDT,M3.2.0,M11.1.0' },
    { text: '(~UTC-08:00) America/Juneau', value: 'America/Juneau', openwrt: 'AKST9AKDT,M3.2.0,M11.1.0' },
    { text: '(~UTC-08:00) America/Metlakatla', value: 'America/Metlakatla' },
    { text: '(~UTC-08:00) America/Nome', value: 'America/Nome', openwrt: 'AKST9AKDT,M3.2.0,M11.1.0' },
    { text: '(~UTC-08:00) America/Sitka', value: 'America/Sitka' },
    { text: '(~UTC-08:00) America/Yakutat', value: 'America/Yakutat', openwrt: 'AKST9AKDT,M3.2.0,M11.1.0' },
    { text: '(~UTC-08:00) Etc/GMT+8', value: 'Etc/GMT+8' },
    { text: '(~UTC-08:00) Pacific/Pitcairn', value: 'Pacific/Pitcairn', openwrt: 'PST8' },
    { text: '(~UTC-08:00) SystemV/PST8', value: 'SystemV/PST8' },
    { text: '(~UTC-08:00) SystemV/PST8PDT', value: 'SystemV/PST8PDT' },
    { text: '(~UTC-08:00) US/Alaska', value: 'US/Alaska' },
    { text: '(~UTC-08:00) AST', value: 'AST' },
    { text: '(~UTC-07:00) America/Chihuahua', value: 'America/Chihuahua', openwrt: 'MST7MDT,M4.1.0,M10.5.0' },
    { text: '(~UTC-07:00) America/Creston', value: 'America/Creston' },
    { text: '(~UTC-07:00) America/Dawson', value: 'America/Dawson', openwrt: 'PST8PDT,M3.2.0,M11.1.0' },
    { text: '(~UTC-07:00) America/Dawson_Creek', value: 'America/Dawson_Creek', openwrt: 'MST7' },
    { text: '(~UTC-07:00) America/Ensenada', value: 'America/Ensenada' },
    { text: '(~UTC-07:00) America/Fort_Nelson', value: 'America/Fort_Nelson' },
    { text: '(~UTC-07:00) America/Hermosillo', value: 'America/Hermosillo', openwrt: 'MST7' },
    { text: '(~UTC-07:00) America/Los_Angeles', value: 'America/Los_Angeles', openwrt: 'PST8PDT,M3.2.0,M11.1.0' },
    { text: '(~UTC-07:00) America/Mazatlan', value: 'America/Mazatlan', openwrt: 'MST7MDT,M4.1.0,M10.5.0' },
    { text: '(~UTC-07:00) America/Phoenix', value: 'America/Phoenix', openwrt: 'MST7' },
    {
      text: '(~UTC-07:00) America/Santa_Isabel',
      value: 'America/Santa_Isabel',
      openwrt: 'PST8PDT,M4.1.0,M10.5.0',
    },
    { text: '(~UTC-07:00) America/Tijuana', value: 'America/Tijuana', openwrt: 'PST8PDT,M3.2.0,M11.1.0' },
    { text: '(~UTC-07:00) America/Vancouver', value: 'America/Vancouver', openwrt: 'PST8PDT,M3.2.0,M11.1.0' },
    { text: '(~UTC-07:00) America/Whitehorse', value: 'America/Whitehorse', openwrt: 'PST8PDT,M3.2.0,M11.1.0' },
    { text: '(~UTC-07:00) Canada/Pacific', value: 'Canada/Pacific' },
    { text: '(~UTC-07:00) Canada/Yukon', value: 'Canada/Yukon' },
    { text: '(~UTC-07:00) Etc/GMT+7', value: 'Etc/GMT+7' },
    { text: '(~UTC-07:00) Mexico/BajaNorte', value: 'Mexico/BajaNorte' },
    { text: '(~UTC-07:00) Mexico/BajaSur', value: 'Mexico/BajaSur' },
    { text: '(~UTC-07:00) PST8PDT', value: 'PST8PDT' },
    { text: '(~UTC-07:00) SystemV/MST7', value: 'SystemV/MST7' },
    { text: '(~UTC-07:00) SystemV/MST7MDT', value: 'SystemV/MST7MDT' },
    { text: '(~UTC-07:00) US/Arizona', value: 'US/Arizona' },
    { text: '(~UTC-07:00) US/Pacific', value: 'US/Pacific' },
    { text: '(~UTC-07:00) US/Pacific-New', value: 'US/Pacific-New' },
    { text: '(~UTC-07:00) MST', value: 'MST' },
    { text: '(~UTC-07:00) PNT', value: 'PNT' },
    { text: '(~UTC-07:00) PST', value: 'PST' },
    { text: '(~UTC-06:00) America/Bahia_Banderas', value: 'America/Bahia_Banderas' },
    { text: '(~UTC-06:00) America/Belize', value: 'America/Belize', openwrt: 'CST6' },
    { text: '(~UTC-06:00) America/Boise', value: 'America/Boise', openwrt: 'MST7MDT,M3.2.0,M11.1.0' },
    {
      text: '(~UTC-06:00) America/Cambridge_Bay',
      value: 'America/Cambridge_Bay',
      openwrt: 'MST7MDT,M3.2.0,M11.1.0',
    },
    { text: '(~UTC-06:00) America/Costa_Rica', value: 'America/Costa_Rica', openwrt: 'CST6' },
    { text: '(~UTC-06:00) America/Denver', value: 'America/Denver', openwrt: 'MST7MDT,M3.2.0,M11.1.0' },
    { text: '(~UTC-06:00) America/Edmonton', value: 'America/Edmonton', openwrt: 'MST7MDT,M3.2.0,M11.1.0' },
    { text: '(~UTC-06:00) America/El_Salvador', value: 'America/El_Salvador', openwrt: 'CST6' },
    { text: '(~UTC-06:00) America/Guatemala', value: 'America/Guatemala', openwrt: 'CST6' },
    { text: '(~UTC-06:00) America/Inuvik', value: 'America/Inuvik', openwrt: 'MST7MDT,M3.2.0,M11.1.0' },
    { text: '(~UTC-06:00) America/Managua', value: 'America/Managua', openwrt: 'CST6' },
    { text: '(~UTC-06:00) America/Merida', value: 'America/Merida', openwrt: 'CST6CDT,M4.1.0,M10.5.0' },
    { text: '(~UTC-06:00) America/Mexico_City', value: 'America/Mexico_City', openwrt: 'CST6CDT,M4.1.0,M10.5.0' },
    { text: '(~UTC-06:00) America/Monterrey', value: 'America/Monterrey', openwrt: 'CST6CDT,M4.1.0,M10.5.0' },
    { text: '(~UTC-06:00) America/Ojinaga', value: 'America/Ojinaga', openwrt: 'MST7MDT,M3.2.0,M11.1.0' },
    { text: '(~UTC-06:00) America/Regina', value: 'America/Regina', openwrt: 'CST6' },
    { text: '(~UTC-06:00) America/Shiprock', value: 'America/Shiprock', openwrt: 'MST7MDT,M3.2.0,M11.1.0' },
    { text: '(~UTC-06:00) America/Swift_Current', value: 'America/Swift_Current', openwrt: 'CST6' },
    { text: '(~UTC-06:00) America/Tegucigalpa', value: 'America/Tegucigalpa', openwrt: 'CST6' },
    { text: '(~UTC-06:00) America/Yellowknife', value: 'America/Yellowknife', openwrt: 'MST7MDT,M3.2.0,M11.1.0' },
    { text: '(~UTC-06:00) Canada/Mountain', value: 'Canada/Mountain' },
    { text: '(~UTC-06:00) Canada/Saskatchewan', value: 'Canada/Saskatchewan' },
    { text: '(~UTC-06:00) Etc/GMT+6', value: 'Etc/GMT+6' },
    { text: '(~UTC-06:00) MST7MDT', value: 'MST7MDT' },
    { text: '(~UTC-06:00) Mexico/General', value: 'Mexico/General' },
    { text: '(~UTC-06:00) Navajo', value: 'Navajo' },
    { text: '(~UTC-06:00) Pacific/Galapagos', value: 'Pacific/Galapagos', openwrt: 'GALT6' },
    { text: '(~UTC-06:00) SystemV/CST6', value: 'SystemV/CST6' },
    { text: '(~UTC-06:00) SystemV/CST6CDT', value: 'SystemV/CST6CDT' },
    { text: '(~UTC-06:00) US/Mountain', value: 'US/Mountain' },
    { text: '(~UTC-05:00) America/Atikokan', value: 'America/Atikokan', openwrt: 'EST5' },
    { text: '(~UTC-05:00) America/Bogota', value: 'America/Bogota', openwrt: 'COT5' },
    { text: '(~UTC-05:00) America/Cancun', value: 'America/Cancun', openwrt: 'CST6CDT,M4.1.0,M10.5.0' },
    { text: '(~UTC-05:00) America/Cayman', value: 'America/Cayman', openwrt: 'EST5' },
    { text: '(~UTC-05:00) America/Chicago', value: 'America/Chicago', openwrt: 'CST6CDT,M3.2.0,M11.1.0' },
    { text: '(~UTC-05:00) America/Coral_Harbour', value: 'America/Coral_Harbour' },
    { text: '(~UTC-05:00) America/Eirunepe', value: 'America/Eirunepe', openwrt: 'AMT4' },
    { text: '(~UTC-05:00) America/Guayaquil', value: 'America/Guayaquil', openwrt: 'ECT5' },
    {
      text: '(~UTC-05:00) America/Indiana/Knox',
      value: 'America/Indiana/Knox',
      openwrt: 'CST6CDT,M3.2.0,M11.1.0',
    },
    {
      text: '(~UTC-05:00) America/Indiana/Tell_City',
      value: 'America/Indiana/Tell_City',
      openwrt: 'CST6CDT,M3.2.0,M11.1.0',
    },
    { text: '(~UTC-05:00) America/Jamaica', value: 'America/Jamaica', openwrt: 'EST5' },
    { text: '(~UTC-05:00) America/Knox_IN', value: 'America/Knox_IN' },
    { text: '(~UTC-05:00) America/Lima', value: 'America/Lima', openwrt: 'PET5' },
    { text: '(~UTC-05:00) America/Matamoros', value: 'America/Matamoros', openwrt: 'CST6CDT,M3.2.0,M11.1.0' },
    { text: '(~UTC-05:00) America/Menominee', value: 'America/Menominee', openwrt: 'CST6CDT,M3.2.0,M11.1.0' },
    { text: '(~UTC-05:00) America/North_Dakota/Beulah', value: 'America/North_Dakota/Beulah' },
    {
      text: '(~UTC-05:00) America/North_Dakota/Center',
      value: 'America/North_Dakota/Center',
      openwrt: 'CST6CDT,M3.2.0,M11.1.0',
    },
    {
      text: '(~UTC-05:00) America/North_Dakota/New_Salem',
      value: 'America/North_Dakota/New_Salem',
      openwrt: 'CST6CDT,M3.2.0,M11.1.0',
    },
    { text: '(~UTC-05:00) America/Panama', value: 'America/Panama', openwrt: 'EST5' },
    { text: '(~UTC-05:00) America/Porto_Acre', value: 'America/Porto_Acre' },
    { text: '(~UTC-05:00) America/Rainy_River', value: 'America/Rainy_River', openwrt: 'CST6CDT,M3.2.0,M11.1.0' },
    {
      text: '(~UTC-05:00) America/Rankin_Inlet',
      value: 'America/Rankin_Inlet',
      openwrt: 'CST6CDT,M3.2.0,M11.1.0',
    },
    { text: '(~UTC-05:00) America/Resolute', value: 'America/Resolute' },
    { text: '(~UTC-05:00) America/Rio_Branco', value: 'America/Rio_Branco', openwrt: 'AMT4' },
    { text: '(~UTC-05:00) America/Winnipeg', value: 'America/Winnipeg', openwrt: 'CST6CDT,M3.2.0,M11.1.0' },
    { text: '(~UTC-05:00) Brazil/Acre', value: 'Brazil/Acre' },
    { text: '(~UTC-05:00) CST6CDT', value: 'CST6CDT' },
    { text: '(~UTC-05:00) Canada/Central', value: 'Canada/Central' },
    { text: '(~UTC-05:00) Chile/EasterIsland', value: 'Chile/EasterIsland' },
    { text: '(~UTC-05:00) Etc/GMT+5', value: 'Etc/GMT+5' },
    { text: '(~UTC-05:00) Jamaica', value: 'Jamaica' },
    { text: '(~UTC-05:00) Pacific/Easter', value: 'Pacific/Easter' },
    { text: '(~UTC-05:00) SystemV/EST5', value: 'SystemV/EST5' },
    { text: '(~UTC-05:00) SystemV/EST5EDT', value: 'SystemV/EST5EDT' },
    { text: '(~UTC-05:00) US/Central', value: 'US/Central' },
    { text: '(~UTC-05:00) US/Indiana-Starke', value: 'US/Indiana-Starke' },
    { text: '(~UTC-05:00) EST', value: 'EST' },
    { text: '(~UTC-05:00) CST', value: 'CST' },
    { text: '(~UTC-04:00) America/Anguilla', value: 'America/Anguilla', openwrt: 'AST4' },
    { text: '(~UTC-04:00) America/Antigua', value: 'America/Antigua', openwrt: 'AST4' },
    { text: '(~UTC-04:00) America/Aruba', value: 'America/Aruba', openwrt: 'AST4' },
    { text: '(~UTC-04:00) America/Asuncion', value: 'America/Asuncion', openwrt: 'PYT4PYST,M10.1.0/0,M4.2.0/0' },
    { text: '(~UTC-04:00) America/Barbados', value: 'America/Barbados', openwrt: 'AST4' },
    { text: '(~UTC-04:00) America/Blanc-Sablon', value: 'America/Blanc-Sablon', openwrt: 'AST4' },
    { text: '(~UTC-04:00) America/Boa_Vista', value: 'America/Boa_Vista', openwrt: 'AMT4' },
    {
      text: '(~UTC-04:00) America/Campo_Grande',
      value: 'America/Campo_Grande',
      openwrt: 'AMT4AMST,M10.3.0/0,M2.3.0/0',
    },
    { text: '(~UTC-04:00) America/Caracas', value: 'America/Caracas', openwrt: 'VET4:30' },
    { text: '(~UTC-04:00) America/Cuiaba', value: 'America/Cuiaba', openwrt: 'AMT4AMST,M10.3.0/0,M2.3.0/0' },
    { text: '(~UTC-04:00) America/Curacao', value: 'America/Curacao', openwrt: 'AST4' },
    { text: '(~UTC-04:00) America/Detroit', value: 'America/Detroit', openwrt: 'EST5EDT,M3.2.0,M11.1.0' },
    { text: '(~UTC-04:00) America/Dominica', value: 'America/Dominica', openwrt: 'AST4' },
    { text: '(~UTC-04:00) America/Fort_Wayne', value: 'America/Fort_Wayne' },
    { text: '(~UTC-04:00) America/Grand_Turk', value: 'America/Grand_Turk', openwrt: 'EST5EDT,M3.2.0,M11.1.0' },
    { text: '(~UTC-04:00) America/Grenada', value: 'America/Grenada', openwrt: 'AST4' },
    { text: '(~UTC-04:00) America/Guadeloupe', value: 'America/Guadeloupe', openwrt: 'AST4' },
    { text: '(~UTC-04:00) America/Guyana', value: 'America/Guyana', openwrt: 'GYT4' },
    { text: '(~UTC-04:00) America/Havana', value: 'America/Havana', openwrt: 'CST5CDT,M3.2.0/0,M10.5.0/1' },
    {
      text: '(~UTC-04:00) America/Indiana/Indianapolis',
      value: 'America/Indiana/Indianapolis',
      openwrt: 'EST5EDT,M3.2.0,M11.1.0',
    },
    {
      text: '(~UTC-04:00) America/Indiana/Marengo',
      value: 'America/Indiana/Marengo',
      openwrt: 'EST5EDT,M3.2.0,M11.1.0',
    },
    {
      text: '(~UTC-04:00) America/Indiana/Petersburg',
      value: 'America/Indiana/Petersburg',
      openwrt: 'EST5EDT,M3.2.0,M11.1.0',
    },
    {
      text: '(~UTC-04:00) America/Indiana/Vevay',
      value: 'America/Indiana/Vevay',
      openwrt: 'EST5EDT,M3.2.0,M11.1.0',
    },
    {
      text: '(~UTC-04:00) America/Indiana/Vincennes',
      value: 'America/Indiana/Vincennes',
      openwrt: 'EST5EDT,M3.2.0,M11.1.0',
    },
    {
      text: '(~UTC-04:00) America/Indiana/Winamac',
      value: 'America/Indiana/Winamac',
      openwrt: 'EST5EDT,M3.2.0,M11.1.0',
    },
    { text: '(~UTC-04:00) America/Indianapolis', value: 'America/Indianapolis' },
    { text: '(~UTC-04:00) America/Iqaluit', value: 'America/Iqaluit', openwrt: 'EST5EDT,M3.2.0,M11.1.0' },
    {
      text: '(~UTC-04:00) America/Kentucky/Louisville',
      value: 'America/Kentucky/Louisville',
      openwrt: 'EST5EDT,M3.2.0,M11.1.0',
    },
    {
      text: '(~UTC-04:00) America/Kentucky/Monticello',
      value: 'America/Kentucky/Monticello',
      openwrt: 'EST5EDT,M3.2.0,M11.1.0',
    },
    { text: '(~UTC-04:00) America/Kralendijk', value: 'America/Kralendijk' },
    { text: '(~UTC-04:00) America/La_Paz', value: 'America/La_Paz', openwrt: 'BOT4' },
    { text: '(~UTC-04:00) America/Louisville', value: 'America/Louisville' },
    { text: '(~UTC-04:00) America/Lower_Princes', value: 'America/Lower_Princes' },
    { text: '(~UTC-04:00) America/Manaus', value: 'America/Manaus', openwrt: 'AMT4' },
    { text: '(~UTC-04:00) America/Marigot', value: 'America/Marigot', openwrt: 'AST4' },
    { text: '(~UTC-04:00) America/Martinique', value: 'America/Martinique', openwrt: 'AST4' },
    { text: '(~UTC-04:00) America/Montreal', value: 'America/Montreal', openwrt: 'EST5EDT,M3.2.0,M11.1.0' },
    { text: '(~UTC-04:00) America/Montserrat', value: 'America/Montserrat', openwrt: 'AST4' },
    { text: '(~UTC-04:00) America/Nassau', value: 'America/Nassau', openwrt: 'EST5EDT,M3.2.0,M11.1.0' },
    { text: '(~UTC-04:00) America/New_York', value: 'America/New_York', openwrt: 'EST5EDT,M3.2.0,M11.1.0' },
    { text: '(~UTC-04:00) America/Nipigon', value: 'America/Nipigon', openwrt: 'EST5EDT,M3.2.0,M11.1.0' },
    { text: '(~UTC-04:00) America/Pangnirtung', value: 'America/Pangnirtung', openwrt: 'EST5EDT,M3.2.0,M11.1.0' },
    { text: '(~UTC-04:00) America/Port-au-Prince', value: 'America/Port-au-Prince', openwrt: 'EST5' },
    { text: '(~UTC-04:00) America/Port_of_Spain', value: 'America/Port_of_Spain', openwrt: 'AST4' },
    { text: '(~UTC-04:00) America/Porto_Velho', value: 'America/Porto_Velho', openwrt: 'AMT4' },
    { text: '(~UTC-04:00) America/Puerto_Rico', value: 'America/Puerto_Rico', openwrt: 'AST4' },
    { text: '(~UTC-04:00) America/Santo_Domingo', value: 'America/Santo_Domingo', openwrt: 'AST4' },
    { text: '(~UTC-04:00) America/St_Barthelemy', value: 'America/St_Barthelemy', openwrt: 'AST4' },
    { text: '(~UTC-04:00) America/St_Kitts', value: 'America/St_Kitts', openwrt: 'AST4' },
    { text: '(~UTC-04:00) America/St_Lucia', value: 'America/St_Lucia', openwrt: 'AST4' },
    { text: '(~UTC-04:00) America/St_Thomas', value: 'America/St_Thomas', openwrt: 'AST4' },
    { text: '(~UTC-04:00) America/St_Vincent', value: 'America/St_Vincent', openwrt: 'AST4' },
    { text: '(~UTC-04:00) America/Thunder_Bay', value: 'America/Thunder_Bay', openwrt: 'EST5EDT,M3.2.0,M11.1.0' },
    { text: '(~UTC-04:00) America/Toronto', value: 'America/Toronto', openwrt: 'EST5EDT,M3.2.0,M11.1.0' },
    { text: '(~UTC-04:00) America/Tortola', value: 'America/Tortola', openwrt: 'AST4' },
    { text: '(~UTC-04:00) America/Virgin', value: 'America/Virgin' },
    { text: '(~UTC-04:00) Brazil/West', value: 'Brazil/West' },
    { text: '(~UTC-04:00) Canada/Eastern', value: 'Canada/Eastern' },
    { text: '(~UTC-04:00) Cuba', value: 'Cuba' },
    { text: '(~UTC-04:00) EST5EDT', value: 'EST5EDT' },
    { text: '(~UTC-04:00) Etc/GMT+4', value: 'Etc/GMT+4' },
    { text: '(~UTC-04:00) SystemV/AST4', value: 'SystemV/AST4' },
    { text: '(~UTC-04:00) SystemV/AST4ADT', value: 'SystemV/AST4ADT' },
    { text: '(~UTC-04:00) US/East-Indiana', value: 'US/East-Indiana' },
    { text: '(~UTC-04:00) US/Eastern', value: 'US/Eastern' },
    { text: '(~UTC-04:00) US/Michigan', value: 'US/Michigan' },
    { text: '(~UTC-04:00) IET', value: 'IET' },
    { text: '(~UTC-04:00) PRT', value: 'PRT' },
    { text: '(~UTC-03:00) America/Araguaina', value: 'America/Araguaina', openwrt: 'BRT3' },
    {
      text: '(~UTC-03:00) America/Argentina/Buenos_Aires',
      value: 'America/Argentina/Buenos_Aires',
      openwrt: 'ART3',
    },
    { text: '(~UTC-03:00) America/Argentina/Catamarca', value: 'America/Argentina/Catamarca', openwrt: 'ART3' },
    { text: '(~UTC-03:00) America/Argentina/ComodRivadavia', value: 'America/Argentina/ComodRivadavia' },
    { text: '(~UTC-03:00) America/Argentina/Cordoba', value: 'America/Argentina/Cordoba', openwrt: 'ART3' },
    { text: '(~UTC-03:00) America/Argentina/Jujuy', value: 'America/Argentina/Jujuy', openwrt: 'ART3' },
    { text: '(~UTC-03:00) America/Argentina/La_Rioja', value: 'America/Argentina/La_Rioja', openwrt: 'ART3' },
    { text: '(~UTC-03:00) America/Argentina/Mendoza', value: 'America/Argentina/Mendoza', openwrt: 'ART3' },
    {
      text: '(~UTC-03:00) America/Argentina/Rio_Gallegos',
      value: 'America/Argentina/Rio_Gallegos',
      openwrt: 'ART3',
    },
    { text: '(~UTC-03:00) America/Argentina/Salta', value: 'America/Argentina/Salta', openwrt: 'ART3' },
    { text: '(~UTC-03:00) America/Argentina/San_Juan', value: 'America/Argentina/San_Juan', openwrt: 'ART3' },
    { text: '(~UTC-03:00) America/Argentina/San_Luis', value: 'America/Argentina/San_Luis' },
    { text: '(~UTC-03:00) America/Argentina/Tucuman', value: 'America/Argentina/Tucuman', openwrt: 'ART3' },
    { text: '(~UTC-03:00) America/Argentina/Ushuaia', value: 'America/Argentina/Ushuaia', openwrt: 'ART3' },
    { text: '(~UTC-03:00) America/Bahia', value: 'America/Bahia', openwrt: 'BRT3' },
    { text: '(~UTC-03:00) America/Belem', value: 'America/Belem', openwrt: 'BRT3' },
    { text: '(~UTC-03:00) America/Buenos_Aires', value: 'America/Buenos_Aires' },
    { text: '(~UTC-03:00) America/Catamarca', value: 'America/Catamarca' },
    { text: '(~UTC-03:00) America/Cayenne', value: 'America/Cayenne', openwrt: 'GFT3' },
    { text: '(~UTC-03:00) America/Cordoba', value: 'America/Cordoba' },
    { text: '(~UTC-03:00) America/Fortaleza', value: 'America/Fortaleza', openwrt: 'BRT3' },
    { text: '(~UTC-03:00) America/Glace_Bay', value: 'America/Glace_Bay', openwrt: 'AST4ADT,M3.2.0,M11.1.0' },
    {
      text: '(~UTC-03:00) America/Goose_Bay',
      value: 'America/Goose_Bay',
      openwrt: 'AST4ADT,M3.2.0/0:01,M11.1.0/0:01',
    },
    { text: '(~UTC-03:00) America/Halifax', value: 'America/Halifax', openwrt: 'AST4ADT,M3.2.0,M11.1.0' },
    { text: '(~UTC-03:00) America/Jujuy', value: 'America/Jujuy' },
    { text: '(~UTC-03:00) America/Maceio', value: 'America/Maceio', openwrt: 'BRT3' },
    { text: '(~UTC-03:00) America/Mendoza', value: 'America/Mendoza' },
    { text: '(~UTC-03:00) America/Moncton', value: 'America/Moncton', openwrt: 'AST4ADT,M3.2.0,M11.1.0' },
    { text: '(~UTC-03:00) America/Montevideo', value: 'America/Montevideo', openwrt: 'UYT3UYST,M10.1.0,M3.2.0' },
    { text: '(~UTC-03:00) America/Paramaribo', value: 'America/Paramaribo', openwrt: 'SRT3' },
    { text: '(~UTC-03:00) America/Punta_Arenas', value: 'America/Punta_Arenas' },
    { text: '(~UTC-03:00) America/Recife', value: 'America/Recife', openwrt: 'BRT3' },
    { text: '(~UTC-03:00) America/Rosario', value: 'America/Rosario' },
    { text: '(~UTC-03:00) America/Santarem', value: 'America/Santarem', openwrt: 'BRT3' },
    { text: '(~UTC-03:00) America/Santiago', value: 'America/Santiago' },
    {
      text: '(~UTC-03:00) America/Sao_Paulo',
      value: 'America/Sao_Paulo',
      openwrt: 'BRT3BRST,M10.3.0/0,M2.3.0/0',
    },
    { text: '(~UTC-03:00) America/Thule', value: 'America/Thule', openwrt: 'AST4ADT,M3.2.0,M11.1.0' },
    { text: '(~UTC-03:00) Antarctica/Palmer', value: 'Antarctica/Palmer' },
    { text: '(~UTC-03:00) Antarctica/Rothera', value: 'Antarctica/Rothera', openwrt: 'ROTT3' },
    { text: '(~UTC-03:00) Atlantic/Bermuda', value: 'Atlantic/Bermuda', openwrt: 'AST4ADT,M3.2.0,M11.1.0' },
    { text: '(~UTC-03:00) Atlantic/Stanley', value: 'Atlantic/Stanley', openwrt: 'FKT4FKST,M9.1.0,M4.3.0' },
    { text: '(~UTC-03:00) Brazil/East', value: 'Brazil/East' },
    { text: '(~UTC-03:00) Canada/Atlantic', value: 'Canada/Atlantic' },
    { text: '(~UTC-03:00) Chile/Continental', value: 'Chile/Continental' },
    { text: '(~UTC-03:00) Etc/GMT+3', value: 'Etc/GMT+3' },
    { text: '(~UTC-03:00) AGT', value: 'AGT' },
    { text: '(~UTC-03:00) BET', value: 'BET' },
    {
      text: '(~UTC-02:30) America/St_Johns',
      value: 'America/St_Johns',
      openwrt: 'NST3:30NDT,M3.2.0/0:01,M11.1.0/0:01',
    },
    { text: '(~UTC-02:30) Canada/Newfoundland', value: 'Canada/Newfoundland' },
    { text: '(~UTC-02:30) CNT', value: 'CNT' },
    { text: '(~UTC-02:00) America/Godthab', value: 'America/Godthab' },
    { text: '(~UTC-02:00) America/Miquelon', value: 'America/Miquelon', openwrt: 'PMST3PMDT,M3.2.0,M11.1.0' },
    { text: '(~UTC-02:00) America/Noronha', value: 'America/Noronha', openwrt: 'FNT2' },
    { text: '(~UTC-02:00) Atlantic/South_Georgia', value: 'Atlantic/South_Georgia', openwrt: 'GST2' },
    { text: '(~UTC-02:00) Brazil/DeNoronha', value: 'Brazil/DeNoronha' },
    { text: '(~UTC-02:00) Etc/GMT+2', value: 'Etc/GMT+2' },
    { text: '(~UTC-01:00) Atlantic/Cape_Verde', value: 'Atlantic/Cape_Verde', openwrt: 'CVT1' },
    { text: '(~UTC-01:00) Etc/GMT+1', value: 'Etc/GMT+1' },
    { text: '(~UTC+00:00) Africa/Abidjan', value: 'Africa/Abidjan', openwrt: 'GMT0' },
    { text: '(~UTC+00:00) Africa/Accra', value: 'Africa/Accra', openwrt: 'GMT0' },
    { text: '(~UTC+00:00) Africa/Bamako', value: 'Africa/Bamako', openwrt: 'GMT0' },
    { text: '(~UTC+00:00) Africa/Banjul', value: 'Africa/Banjul', openwrt: 'GMT0' },
    { text: '(~UTC+00:00) Africa/Bissau', value: 'Africa/Bissau', openwrt: 'GMT0' },
    { text: '(~UTC+00:00) Africa/Conakry', value: 'Africa/Conakry', openwrt: 'GMT0' },
    { text: '(~UTC+00:00) Africa/Dakar', value: 'Africa/Dakar', openwrt: 'GMT0' },
    { text: '(~UTC+00:00) Africa/Freetown', value: 'Africa/Freetown', openwrt: 'GMT0' },
    { text: '(~UTC+00:00) Africa/Lome', value: 'Africa/Lome', openwrt: 'GMT0' },
    { text: '(~UTC+00:00) Africa/Monrovia', value: 'Africa/Monrovia', openwrt: 'GMT0' },
    { text: '(~UTC+00:00) Africa/Nouakchott', value: 'Africa/Nouakchott', openwrt: 'GMT0' },
    { text: '(~UTC+00:00) Africa/Ouagadougou', value: 'Africa/Ouagadougou', openwrt: 'GMT0' },
    { text: '(~UTC+00:00) Africa/Timbuktu', value: 'Africa/Timbuktu' },
    { text: '(~UTC+00:00) America/Danmarkshavn', value: 'America/Danmarkshavn', openwrt: 'GMT0' },
    {
      text: '(~UTC+00:00) America/Scoresbysund',
      value: 'America/Scoresbysund',
      openwrt: 'EGT1EGST,M3.5.0/0,M10.5.0/1',
    },
    { text: '(~UTC+00:00) Atlantic/Azores', value: 'Atlantic/Azores', openwrt: 'AZOT1AZOST,M3.5.0/0,M10.5.0/1' },
    { text: '(~UTC+00:00) Atlantic/Reykjavik', value: 'Atlantic/Reykjavik', openwrt: 'GMT0' },
    { text: '(~UTC+00:00) Atlantic/St_Helena', value: 'Atlantic/St_Helena', openwrt: 'GMT0' },
    { text: '(~UTC+00:00) Etc/GMT', value: 'Etc/GMT' },
    { text: '(~UTC+00:00) Etc/GMT+0', value: 'Etc/GMT+0' },
    { text: '(~UTC+00:00) Etc/GMT-0', value: 'Etc/GMT-0' },
    { text: '(~UTC+00:00) Etc/GMT0', value: 'Etc/GMT0' },
    { text: '(~UTC+00:00) Etc/Greenwich', value: 'Etc/Greenwich' },
    { text: '(~UTC+00:00) Etc/UCT', value: 'Etc/UCT' },
    { text: '(~UTC+00:00) Etc/UTC', value: 'Etc/UTC' },
    { text: '(~UTC+00:00) Etc/Universal', value: 'Etc/Universal' },
    { text: '(~UTC+00:00) Etc/Zulu', value: 'Etc/Zulu' },
    { text: '(~UTC+00:00) GMT', value: 'GMT' },
    { text: '(~UTC+00:00) GMT0', value: 'GMT0' },
    { text: '(~UTC+00:00) Greenwich', value: 'Greenwich' },
    { text: '(~UTC+00:00) Iceland', value: 'Iceland' },
    { text: '(~UTC+00:00) UCT', value: 'UCT' },
    { text: '(~UTC+00:00) UTC', value: 'UTC', openwrt: 'UTC' },
    { text: '(~UTC+00:00) Universal', value: 'Universal' },
    { text: '(~UTC+00:00) Zulu', value: 'Zulu' },
    { text: '(~UTC+01:00) Africa/Algiers', value: 'Africa/Algiers', openwrt: 'CET-1' },
    { text: '(~UTC+01:00) Africa/Bangui', value: 'Africa/Bangui', openwrt: 'WAT-1' },
    { text: '(~UTC+01:00) Africa/Brazzaville', value: 'Africa/Brazzaville', openwrt: 'WAT-1' },
    { text: '(~UTC+01:00) Africa/Casablanca', value: 'Africa/Casablanca', openwrt: 'WET0' },
    { text: '(~UTC+01:00) Africa/Douala', value: 'Africa/Douala', openwrt: 'WAT-1' },
    { text: '(~UTC+01:00) Africa/El_Aaiun', value: 'Africa/El_Aaiun', openwrt: 'WET0' },
    { text: '(~UTC+01:00) Africa/Kinshasa', value: 'Africa/Kinshasa', openwrt: 'WAT-1' },
    { text: '(~UTC+01:00) Africa/Lagos', value: 'Africa/Lagos', openwrt: 'WAT-1' },
    { text: '(~UTC+01:00) Africa/Libreville', value: 'Africa/Libreville', openwrt: 'WAT-1' },
    { text: '(~UTC+01:00) Africa/Luanda', value: 'Africa/Luanda', openwrt: 'WAT-1' },
    { text: '(~UTC+01:00) Africa/Malabo', value: 'Africa/Malabo', openwrt: 'WAT-1' },
    { text: '(~UTC+01:00) Africa/Ndjamena', value: 'Africa/Ndjamena', openwrt: 'WAT-1' },
    { text: '(~UTC+01:00) Africa/Niamey', value: 'Africa/Niamey', openwrt: 'WAT-1' },
    { text: '(~UTC+01:00) Africa/Porto-Novo', value: 'Africa/Porto-Novo', openwrt: 'WAT-1' },
    { text: '(~UTC+01:00) Africa/Sao_Tome', value: 'Africa/Sao_Tome', openwrt: 'GMT0' },
    { text: '(~UTC+01:00) Africa/Tunis', value: 'Africa/Tunis', openwrt: 'CET-1' },
    { text: '(~UTC+01:00) Atlantic/Canary', value: 'Atlantic/Canary', openwrt: 'WET0WEST,M3.5.0/1,M10.5.0' },
    { text: '(~UTC+01:00) Atlantic/Faeroe', value: 'Atlantic/Faeroe' },
    { text: '(~UTC+01:00) Atlantic/Faroe', value: 'Atlantic/Faroe', openwrt: 'WET0WEST,M3.5.0/1,M10.5.0' },
    { text: '(~UTC+01:00) Atlantic/Madeira', value: 'Atlantic/Madeira', openwrt: 'WET0WEST,M3.5.0/1,M10.5.0' },
    { text: '(~UTC+01:00) Eire', value: 'Eire' },
    { text: '(~UTC+01:00) Etc/GMT-1', value: 'Etc/GMT-1' },
    { text: '(~UTC+01:00) Europe/Belfast', value: 'Europe/Belfast' },
    { text: '(~UTC+01:00) Europe/Dublin', value: 'Europe/Dublin', openwrt: 'GMT0IST,M3.5.0/1,M10.5.0' },
    { text: '(~UTC+01:00) Europe/Guernsey', value: 'Europe/Guernsey', openwrt: 'GMT0BST,M3.5.0/1,M10.5.0' },
    { text: '(~UTC+01:00) Europe/Isle_of_Man', value: 'Europe/Isle_of_Man', openwrt: 'GMT0BST,M3.5.0/1,M10.5.0' },
    { text: '(~UTC+01:00) Europe/Jersey', value: 'Europe/Jersey', openwrt: 'GMT0BST,M3.5.0/1,M10.5.0' },
    { text: '(~UTC+01:00) Europe/Lisbon', value: 'Europe/Lisbon', openwrt: 'WET0WEST,M3.5.0/1,M10.5.0' },
    { text: '(~UTC+01:00) Europe/London', value: 'Europe/London', openwrt: 'GMT0BST,M3.5.0/1,M10.5.0' },
    { text: '(~UTC+01:00) GB', value: 'GB' },
    { text: '(~UTC+01:00) GB-Eire', value: 'GB-Eire' },
    { text: '(~UTC+01:00) Portugal', value: 'Portugal' },
    { text: '(~UTC+01:00) WET', value: 'WET' },
    { text: '(~UTC+02:00) Africa/Blantyre', value: 'Africa/Blantyre', openwrt: 'CAT-2' },
    { text: '(~UTC+02:00) Africa/Bujumbura', value: 'Africa/Bujumbura', openwrt: 'CAT-2' },
    { text: '(~UTC+02:00) Africa/Cairo', value: 'Africa/Cairo' },
    { text: '(~UTC+02:00) Africa/Ceuta', value: 'Africa/Ceuta', openwrt: 'CET-1CEST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+02:00) Africa/Gaborone', value: 'Africa/Gaborone', openwrt: 'CAT-2' },
    { text: '(~UTC+02:00) Africa/Harare', value: 'Africa/Harare', openwrt: 'CAT-2' },
    { text: '(~UTC+02:00) Africa/Johannesburg', value: 'Africa/Johannesburg', openwrt: 'SAST-2' },
    { text: '(~UTC+02:00) Africa/Khartoum', value: 'Africa/Khartoum', openwrt: 'EAT-3' },
    { text: '(~UTC+02:00) Africa/Kigali', value: 'Africa/Kigali', openwrt: 'CAT-2' },
    { text: '(~UTC+02:00) Africa/Lubumbashi', value: 'Africa/Lubumbashi', openwrt: 'CAT-2' },
    { text: '(~UTC+02:00) Africa/Lusaka', value: 'Africa/Lusaka', openwrt: 'CAT-2' },
    { text: '(~UTC+02:00) Africa/Maputo', value: 'Africa/Maputo', openwrt: 'CAT-2' },
    { text: '(~UTC+02:00) Africa/Maseru', value: 'Africa/Maseru', openwrt: 'SAST-2' },
    { text: '(~UTC+02:00) Africa/Mbabane', value: 'Africa/Mbabane', openwrt: 'SAST-2' },
    { text: '(~UTC+02:00) Africa/Tripoli', value: 'Africa/Tripoli', openwrt: 'EET-2' },
    { text: '(~UTC+02:00) Africa/Windhoek', value: 'Africa/Windhoek', openwrt: 'WAT-1WAST,M9.1.0,M4.1.0' },
    { text: '(~UTC+02:00) Antarctica/Troll', value: 'Antarctica/Troll' },
    {
      text: '(~UTC+02:00) Arctic/Longyearbyen',
      value: 'Arctic/Longyearbyen',
      openwrt: 'CET-1CEST,M3.5.0,M10.5.0/3',
    },
    { text: '(~UTC+02:00) Atlantic/Jan_Mayen', value: 'Atlantic/Jan_Mayen' },
    { text: '(~UTC+02:00) CET', value: 'CET' },
    { text: '(~UTC+02:00) Egypt', value: 'Egypt' },
    { text: '(~UTC+02:00) Etc/GMT-2', value: 'Etc/GMT-2' },
    { text: '(~UTC+02:00) Europe/Amsterdam', value: 'Europe/Amsterdam', openwrt: 'CET-1CEST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+02:00) Europe/Andorra', value: 'Europe/Andorra', openwrt: 'CET-1CEST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+02:00) Europe/Belgrade', value: 'Europe/Belgrade', openwrt: 'CET-1CEST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+02:00) Europe/Berlin', value: 'Europe/Berlin', openwrt: 'CET-1CEST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+02:00) Europe/Bratislava', value: 'Europe/Bratislava', openwrt: 'CET-1CEST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+02:00) Europe/Brussels', value: 'Europe/Brussels', openwrt: 'CET-1CEST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+02:00) Europe/Budapest', value: 'Europe/Budapest', openwrt: 'CET-1CEST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+02:00) Europe/Busingen', value: 'Europe/Busingen' },
    { text: '(~UTC+02:00) Europe/Copenhagen', value: 'Europe/Copenhagen', openwrt: 'CET-1CEST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+02:00) Europe/Gibraltar', value: 'Europe/Gibraltar', openwrt: 'CET-1CEST,M3.5.0,M10.5.0/3' },
    {
      text: '(~UTC+02:00) Europe/Kaliningrad',
      value: 'Europe/Kaliningrad',
      openwrt: 'EET-2EEST,M3.5.0,M10.5.0/3',
    },
    { text: '(~UTC+02:00) Europe/Ljubljana', value: 'Europe/Ljubljana', openwrt: 'CET-1CEST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+02:00) Europe/Luxembourg', value: 'Europe/Luxembourg', openwrt: 'CET-1CEST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+02:00) Europe/Madrid', value: 'Europe/Madrid', openwrt: 'CET-1CEST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+02:00) Europe/Malta', value: 'Europe/Malta', openwrt: 'CET-1CEST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+02:00) Europe/Monaco', value: 'Europe/Monaco', openwrt: 'CET-1CEST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+02:00) Europe/Oslo', value: 'Europe/Oslo', openwrt: 'CET-1CEST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+02:00) Europe/Paris', value: 'Europe/Paris', openwrt: 'CET-1CEST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+02:00) Europe/Podgorica', value: 'Europe/Podgorica', openwrt: 'CET-1CEST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+02:00) Europe/Prague', value: 'Europe/Prague', openwrt: 'CET-1CEST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+02:00) Europe/Rome', value: 'Europe/Rome', openwrt: 'CET-1CEST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+02:00) Europe/San_Marino', value: 'Europe/San_Marino', openwrt: 'CET-1CEST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+02:00) Europe/Sarajevo', value: 'Europe/Sarajevo', openwrt: 'CET-1CEST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+02:00) Europe/Skopje', value: 'Europe/Skopje', openwrt: 'CET-1CEST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+02:00) Europe/Stockholm', value: 'Europe/Stockholm', openwrt: 'CET-1CEST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+02:00) Europe/Tirane', value: 'Europe/Tirane', openwrt: 'CET-1CEST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+02:00) Europe/Vaduz', value: 'Europe/Vaduz', openwrt: 'CET-1CEST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+02:00) Europe/Vatican', value: 'Europe/Vatican', openwrt: 'CET-1CEST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+02:00) Europe/Vienna', value: 'Europe/Vienna', openwrt: 'CET-1CEST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+02:00) Europe/Warsaw', value: 'Europe/Warsaw', openwrt: 'CET-1CEST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+02:00) Europe/Zagreb', value: 'Europe/Zagreb', openwrt: 'CET-1CEST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+02:00) Europe/Zurich', value: 'Europe/Zurich', openwrt: 'CET-1CEST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+02:00) Libya', value: 'Libya' },
    { text: '(~UTC+02:00) MET', value: 'MET' },
    { text: '(~UTC+02:00) Poland', value: 'Poland' },
    { text: '(~UTC+02:00) ART', value: 'ART' },
    { text: '(~UTC+02:00) CAT', value: 'CAT' },
    { text: '(~UTC+02:00) ECT', value: 'ECT' },
    { text: '(~UTC+03:00) Africa/Addis_Ababa', value: 'Africa/Addis_Ababa' },
    { text: '(~UTC+03:00) Africa/Asmara', value: 'Africa/Asmara', openwrt: 'EAT-3' },
    { text: '(~UTC+03:00) Africa/Asmera', value: 'Africa/Asmera' },
    { text: '(~UTC+03:00) Africa/Dar_es_Salaam', value: 'Africa/Dar_es_Salaam', openwrt: 'EAT-3' },
    { text: '(~UTC+03:00) Africa/Djibouti', value: 'Africa/Djibouti', openwrt: 'EAT-3' },
    { text: '(~UTC+03:00) Africa/Juba', value: 'Africa/Juba' },
    { text: '(~UTC+03:00) Africa/Kampala', value: 'Africa/Kampala', openwrt: 'EAT-3' },
    { text: '(~UTC+03:00) Africa/Mogadishu', value: 'Africa/Mogadishu', openwrt: 'EAT-3' },
    { text: '(~UTC+03:00) Africa/Nairobi', value: 'Africa/Nairobi', openwrt: 'EAT-3' },
    { text: '(~UTC+03:00) Antarctica/Syowa', value: 'Antarctica/Syowa', openwrt: 'SYOT-3' },
    { text: '(~UTC+03:00) Asia/Aden', value: 'Asia/Aden', openwrt: 'AST-3' },
    { text: '(~UTC+03:00) Asia/Amman', value: 'Asia/Amman' },
    { text: '(~UTC+03:00) Asia/Baghdad', value: 'Asia/Baghdad', openwrt: 'AST-3' },
    { text: '(~UTC+03:00) Asia/Bahrain', value: 'Asia/Bahrain', openwrt: 'AST-3' },
    { text: '(~UTC+03:00) Asia/Beirut', value: 'Asia/Beirut', openwrt: 'EET-2EEST,M3.5.0/0,M10.5.0/0' },
    { text: '(~UTC+03:00) Asia/Damascus', value: 'Asia/Damascus', openwrt: 'EET-2EEST,M4.1.5/0,M10.5.5/0' },
    { text: '(~UTC+03:00) Asia/Famagusta', value: 'Asia/Famagusta' },
    { text: '(~UTC+03:00) Asia/Gaza', value: 'Asia/Gaza', openwrt: 'EET-2EEST,M3.5.6/0:01,M9.1.5' },
    { text: '(~UTC+03:00) Asia/Hebron', value: 'Asia/Hebron' },
    { text: '(~UTC+03:00) Asia/Istanbul', value: 'Asia/Istanbul' },
    { text: '(~UTC+03:00) Asia/Jerusalem', value: 'Asia/Jerusalem' },
    { text: '(~UTC+03:00) Asia/Kuwait', value: 'Asia/Kuwait', openwrt: 'AST-3' },
    { text: '(~UTC+03:00) Asia/Nicosia', value: 'Asia/Nicosia', openwrt: 'EET-2EEST,M3.5.0/3,M10.5.0/4' },
    { text: '(~UTC+03:00) Asia/Qatar', value: 'Asia/Qatar', openwrt: 'AST-3' },
    { text: '(~UTC+03:00) Asia/Riyadh', value: 'Asia/Riyadh', openwrt: 'AST-3' },
    { text: '(~UTC+03:00) Asia/Tel_Aviv', value: 'Asia/Tel_Aviv' },
    { text: '(~UTC+03:00) EET', value: 'EET' },
    { text: '(~UTC+03:00) Etc/GMT-3', value: 'Etc/GMT-3' },
    { text: '(~UTC+03:00) Europe/Athens', value: 'Europe/Athens', openwrt: 'EET-2EEST,M3.5.0/3,M10.5.0/4' },
    { text: '(~UTC+03:00) Europe/Bucharest', value: 'Europe/Bucharest', openwrt: 'EET-2EEST,M3.5.0/3,M10.5.0/4' },
    { text: '(~UTC+03:00) Europe/Chisinau', value: 'Europe/Chisinau', openwrt: 'EET-2EEST,M3.5.0/3,M10.5.0/4' },
    { text: '(~UTC+03:00) Europe/Helsinki', value: 'Europe/Helsinki', openwrt: 'EET-2EEST,M3.5.0/3,M10.5.0/4' },
    { text: '(~UTC+03:00) Europe/Istanbul', value: 'Europe/Istanbul', openwrt: 'EET-2EEST,M3.5.0/3,M10.5.0/4' },
    { text: '(~UTC+03:00) Europe/Kiev', value: 'Europe/Kiev', openwrt: 'EET-2EEST,M3.5.0/3,M10.5.0/4' },
    { text: '(~UTC+03:00) Europe/Kirov', value: 'Europe/Kirov' },
    { text: '(~UTC+03:00) Europe/Mariehamn', value: 'Europe/Mariehamn', openwrt: 'EET-2EEST,M3.5.0/3,M10.5.0/4' },
    { text: '(~UTC+03:00) Europe/Minsk', value: 'Europe/Minsk', openwrt: 'EET-2EEST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+03:00) Europe/Moscow', value: 'Europe/Moscow', openwrt: 'MSK-4' },
    { text: '(~UTC+03:00) Europe/Nicosia', value: 'Europe/Nicosia' },
    { text: '(~UTC+03:00) Europe/Riga', value: 'Europe/Riga', openwrt: 'EET-2EEST,M3.5.0/3,M10.5.0/4' },
    {
      text: '(~UTC+03:00) Europe/Simferopol',
      value: 'Europe/Simferopol',
      openwrt: 'EET-2EEST,M3.5.0/3,M10.5.0/4',
    },
    { text: '(~UTC+03:00) Europe/Sofia', value: 'Europe/Sofia', openwrt: 'EET-2EEST,M3.5.0/3,M10.5.0/4' },
    { text: '(~UTC+03:00) Europe/Tallinn', value: 'Europe/Tallinn', openwrt: 'EET-2EEST,M3.5.0/3,M10.5.0/4' },
    { text: '(~UTC+03:00) Europe/Tiraspol', value: 'Europe/Tiraspol' },
    { text: '(~UTC+03:00) Europe/Uzhgorod', value: 'Europe/Uzhgorod', openwrt: 'EET-2EEST,M3.5.0/3,M10.5.0/4' },
    { text: '(~UTC+03:00) Europe/Vilnius', value: 'Europe/Vilnius', openwrt: 'EET-2EEST,M3.5.0/3,M10.5.0/4' },
    {
      text: '(~UTC+03:00) Europe/Zaporozhye',
      value: 'Europe/Zaporozhye',
      openwrt: 'EET-2EEST,M3.5.0/3,M10.5.0/4',
    },
    { text: '(~UTC+03:00) Indian/Antananarivo', value: 'Indian/Antananarivo', openwrt: 'EAT-3' },
    { text: '(~UTC+03:00) Indian/Comoro', value: 'Indian/Comoro', openwrt: 'EAT-3' },
    { text: '(~UTC+03:00) Indian/Mayotte', value: 'Indian/Mayotte', openwrt: 'EAT-3' },
    { text: '(~UTC+03:00) Israel', value: 'Israel' },
    { text: '(~UTC+03:00) Turkey', value: 'Turkey' },
    { text: '(~UTC+03:00) W-SU', value: 'W-SU' },
    { text: '(~UTC+03:00) EAT', value: 'EAT' },
    { text: '(~UTC+04:00) Asia/Baku', value: 'Asia/Baku', openwrt: 'AZT-4AZST,M3.5.0/4,M10.5.0/5' },
    { text: '(~UTC+04:00) Asia/Dubai', value: 'Asia/Dubai', openwrt: 'GST-4' },
    { text: '(~UTC+04:00) Asia/Muscat', value: 'Asia/Muscat', openwrt: 'GST-4' },
    { text: '(~UTC+04:00) Asia/Tbilisi', value: 'Asia/Tbilisi', openwrt: 'GET-4' },
    { text: '(~UTC+04:00) Asia/Yerevan', value: 'Asia/Yerevan', openwrt: 'AMT-4AMST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+04:00) Etc/GMT-4', value: 'Etc/GMT-4' },
    { text: '(~UTC+04:00) Europe/Astrakhan', value: 'Europe/Astrakhan' },
    { text: '(~UTC+04:00) Europe/Samara', value: 'Europe/Samara', openwrt: 'SAMT-3SAMST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+04:00) Europe/Saratov', value: 'Europe/Saratov' },
    { text: '(~UTC+04:00) Europe/Ulyanovsk', value: 'Europe/Ulyanovsk' },
    { text: '(~UTC+04:00) Europe/Volgograd', value: 'Europe/Volgograd', openwrt: 'VOLT-3VOLST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+04:00) Indian/Mahe', value: 'Indian/Mahe', openwrt: 'SCT-4' },
    { text: '(~UTC+04:00) Indian/Mauritius', value: 'Indian/Mauritius', openwrt: 'MUT-4' },
    { text: '(~UTC+04:00) Indian/Reunion', value: 'Indian/Reunion', openwrt: 'RET-4' },
    { text: '(~UTC+04:00) NET', value: 'NET' },
    { text: '(~UTC+04:30) Asia/Kabul', value: 'Asia/Kabul', openwrt: 'AFT-4:30' },
    { text: '(~UTC+04:30) Asia/Tehran', value: 'Asia/Tehran', openwrt: 'IRST-3:30IRDT,80/0,264/0' },
    { text: '(~UTC+04:30) Iran', value: 'Iran' },
    { text: '(~UTC+05:00) Antarctica/Mawson', value: 'Antarctica/Mawson', openwrt: 'MAWT-5' },
    { text: '(~UTC+05:00) Asia/Aqtau', value: 'Asia/Aqtau', openwrt: 'AQTT-5' },
    { text: '(~UTC+05:00) Asia/Aqtobe', value: 'Asia/Aqtobe', openwrt: 'AQTT-5' },
    { text: '(~UTC+05:00) Asia/Ashgabat', value: 'Asia/Ashgabat', openwrt: 'TMT-5' },
    { text: '(~UTC+05:00) Asia/Ashkhabad', value: 'Asia/Ashkhabad' },
    { text: '(~UTC+05:00) Asia/Atyrau', value: 'Asia/Atyrau' },
    { text: '(~UTC+05:00) Asia/Dushanbe', value: 'Asia/Dushanbe', openwrt: 'TJT-5' },
    { text: '(~UTC+05:00) Asia/Karachi', value: 'Asia/Karachi', openwrt: 'PKT-5' },
    { text: '(~UTC+05:00) Asia/Oral', value: 'Asia/Oral', openwrt: 'ORAT-5' },
    { text: '(~UTC+05:00) Asia/Samarkand', value: 'Asia/Samarkand', openwrt: 'UZT-5' },
    { text: '(~UTC+05:00) Asia/Tashkent', value: 'Asia/Tashkent', openwrt: 'UZT-5' },
    {
      text: '(~UTC+05:00) Asia/Yekaterinburg',
      value: 'Asia/Yekaterinburg',
      openwrt: 'YEKT-5YEKST,M3.5.0,M10.5.0/3',
    },
    { text: '(~UTC+05:00) Etc/GMT-5', value: 'Etc/GMT-5' },
    { text: '(~UTC+05:00) Indian/Kerguelen', value: 'Indian/Kerguelen', openwrt: 'TFT-5' },
    { text: '(~UTC+05:00) Indian/Maldives', value: 'Indian/Maldives', openwrt: 'MVT-5' },
    { text: '(~UTC+05:00) PLT', value: 'PLT' },
    { text: '(~UTC+05:30) Asia/Calcutta', value: 'Asia/Calcutta' },
    { text: '(~UTC+05:30) Asia/Colombo', value: 'Asia/Colombo', openwrt: 'IST-5:30' },
    { text: '(~UTC+05:30) Asia/Kolkata', value: 'Asia/Kolkata', openwrt: 'IST-5:30' },
    { text: '(~UTC+05:30) IST', value: 'IST' },
    { text: '(~UTC+05:45) Asia/Kathmandu', value: 'Asia/Kathmandu', openwrt: 'NPT-5:45' },
    { text: '(~UTC+05:45) Asia/Katmandu', value: 'Asia/Katmandu' },
    { text: '(~UTC+06:00) Antarctica/Vostok', value: 'Antarctica/Vostok', openwrt: 'VOST-6' },
    { text: '(~UTC+06:00) Asia/Almaty', value: 'Asia/Almaty', openwrt: 'ALMT-6' },
    { text: '(~UTC+06:00) Asia/Bishkek', value: 'Asia/Bishkek', openwrt: 'KGT-6' },
    { text: '(~UTC+06:00) Asia/Dacca', value: 'Asia/Dacca' },
    { text: '(~UTC+06:00) Asia/Dhaka', value: 'Asia/Dhaka', openwrt: 'BDT-6' },
    { text: '(~UTC+06:00) Asia/Kashgar', value: 'Asia/Kashgar', openwrt: 'CST-8' },
    { text: '(~UTC+06:00) Asia/Omsk', value: 'Asia/Omsk', openwrt: 'OMST-7' },
    { text: '(~UTC+06:00) Asia/Qyzylorda', value: 'Asia/Qyzylorda', openwrt: 'QYZT-6' },
    { text: '(~UTC+06:00) Asia/Thimbu', value: 'Asia/Thimbu' },
    { text: '(~UTC+06:00) Asia/Thimphu', value: 'Asia/Thimphu', openwrt: 'BTT-6' },
    { text: '(~UTC+06:00) Asia/Urumqi', value: 'Asia/Urumqi', openwrt: 'CST-8' },
    { text: '(~UTC+06:00) Etc/GMT-6', value: 'Etc/GMT-6' },
    { text: '(~UTC+06:00) Indian/Chagos', value: 'Indian/Chagos', openwrt: 'IOT-6' },
    { text: '(~UTC+06:00) BST', value: 'BST' },
    { text: '(~UTC+06:30) Asia/Rangoon', value: 'Asia/Rangoon', openwrt: 'MMT-6:30' },
    { text: '(~UTC+06:30) Asia/Yangon', value: 'Asia/Yangon' },
    { text: '(~UTC+06:30) Indian/Cocos', value: 'Indian/Cocos', openwrt: 'CCT-6:30' },
    { text: '(~UTC+07:00) Antarctica/Davis', value: 'Antarctica/Davis', openwrt: 'DAVT-7' },
    { text: '(~UTC+07:00) Asia/Bangkok', value: 'Asia/Bangkok', openwrt: 'ICT-7' },
    { text: '(~UTC+07:00) Asia/Barnaul', value: 'Asia/Barnaul' },
    { text: '(~UTC+07:00) Asia/Ho_Chi_Minh', value: 'Asia/Ho_Chi_Minh', openwrt: 'ICT-7' },
    { text: '(~UTC+07:00) Asia/Hovd', value: 'Asia/Hovd', openwrt: 'HOVT-7' },
    { text: '(~UTC+07:00) Asia/Jakarta', value: 'Asia/Jakarta', openwrt: 'WIT-7' },
    { text: '(~UTC+07:00) Asia/Krasnoyarsk', value: 'Asia/Krasnoyarsk', openwrt: 'KRAT-7KRAST,M3.5.0,M10.5.0/3' },
    {
      text: '(~UTC+07:00) Asia/Novokuznetsk',
      value: 'Asia/Novokuznetsk',
      openwrt: 'NOVT-6NOVST,M3.5.0,M10.5.0/3',
    },
    { text: '(~UTC+07:00) Asia/Novosibirsk', value: 'Asia/Novosibirsk', openwrt: 'NOVT-6NOVST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+07:00) Asia/Phnom_Penh', value: 'Asia/Phnom_Penh', openwrt: 'ICT-7' },
    { text: '(~UTC+07:00) Asia/Pontianak', value: 'Asia/Pontianak', openwrt: 'WIT-7' },
    { text: '(~UTC+07:00) Asia/Saigon', value: 'Asia/Saigon' },
    { text: '(~UTC+07:00) Asia/Tomsk', value: 'Asia/Tomsk' },
    { text: '(~UTC+07:00) Asia/Vientiane', value: 'Asia/Vientiane', openwrt: 'ICT-7' },
    { text: '(~UTC+07:00) Etc/GMT-7', value: 'Etc/GMT-7' },
    { text: '(~UTC+07:00) Indian/Christmas', value: 'Indian/Christmas', openwrt: 'CXT-7' },
    { text: '(~UTC+07:00) VST', value: 'VST' },
    { text: '(~UTC+08:00) Antarctica/Casey', value: 'Antarctica/Casey', openwrt: 'WST-8' },
    { text: '(~UTC+08:00) Asia/Brunei', value: 'Asia/Brunei', openwrt: 'BNT-8' },
    { text: '(~UTC+08:00) Asia/Choibalsan', value: 'Asia/Choibalsan', openwrt: 'CHOT-8' },
    { text: '(~UTC+08:00) Asia/Chongqing', value: 'Asia/Chongqing', openwrt: 'CST-8' },
    { text: '(~UTC+08:00) Asia/Chungking', value: 'Asia/Chungking' },
    { text: '(~UTC+08:00) Asia/Harbin', value: 'Asia/Harbin', openwrt: 'CST-8' },
    { text: '(~UTC+08:00) Asia/Hong_Kong', value: 'Asia/Hong_Kong' },
    { text: '(~UTC+08:00) Asia/Irkutsk', value: 'Asia/Irkutsk', openwrt: 'IRKT-8IRKST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+08:00) Asia/Kuala_Lumpur', value: 'Asia/Kuala_Lumpur', openwrt: 'MYT-8' },
    { text: '(~UTC+08:00) Asia/Kuching', value: 'Asia/Kuching', openwrt: 'MYT-8' },
    { text: '(~UTC+08:00) Asia/Macao', value: 'Asia/Macao' },
    { text: '(~UTC+08:00) Asia/Macau', value: 'Asia/Macau', openwrt: 'CST-8' },
    { text: '(~UTC+08:00) Asia/Makassar', value: 'Asia/Makassar', openwrt: 'CIT-8' },
    { text: '(~UTC+08:00) Asia/Manila', value: 'Asia/Manila', openwrt: 'PHT-8' },
    { text: '(~UTC+08:00) Asia/Shanghai', value: 'Asia/Shanghai', openwrt: 'CST-8' },
    { text: '(~UTC+08:00) Asia/Singapore', value: 'Asia/Singapore', openwrt: 'SGT-8' },
    { text: '(~UTC+08:00) Asia/Taipei', value: 'Asia/Taipei', openwrt: 'CST-8' },
    { text: '(~UTC+08:00) Asia/Ujung_Pandang', value: 'Asia/Ujung_Pandang' },
    { text: '(~UTC+08:00) Asia/Ulaanbaatar', value: 'Asia/Ulaanbaatar', openwrt: 'ULAT-8' },
    { text: '(~UTC+08:00) Asia/Ulan_Bator', value: 'Asia/Ulan_Bator' },
    { text: '(~UTC+08:00) Australia/Perth', value: 'Australia/Perth', openwrt: 'WST-8' },
    { text: '(~UTC+08:00) Australia/West', value: 'Australia/West' },
    { text: '(~UTC+08:00) Etc/GMT-8', value: 'Etc/GMT-8' },
    { text: '(~UTC+08:00) Hongkong', value: 'Hongkong' },
    { text: '(~UTC+08:00) PRC', value: 'PRC' },
    { text: '(~UTC+08:00) Singapore', value: 'Singapore' },
    { text: '(~UTC+08:00) CTT', value: 'CTT' },
    { text: '(~UTC+08:45) Australia/Eucla', value: 'Australia/Eucla', openwrt: 'CWST-8:45' },
    { text: '(~UTC+09:00) Asia/Chita', value: 'Asia/Chita' },
    { text: '(~UTC+09:00) Asia/Dili', value: 'Asia/Dili', openwrt: 'TLT-9' },
    { text: '(~UTC+09:00) Asia/Jayapura', value: 'Asia/Jayapura', openwrt: 'EIT-9' },
    { text: '(~UTC+09:00) Asia/Khandyga', value: 'Asia/Khandyga' },
    { text: '(~UTC+09:00) Asia/Pyongyang', value: 'Asia/Pyongyang', openwrt: 'KST-9' },
    { text: '(~UTC+09:00) Asia/Seoul', value: 'Asia/Seoul', openwrt: 'KST-9' },
    { text: '(~UTC+09:00) Asia/Tokyo', value: 'Asia/Tokyo', openwrt: 'JST-9' },
    { text: '(~UTC+09:00) Asia/Yakutsk', value: 'Asia/Yakutsk', openwrt: 'YAKT-9YAKST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+09:00) Etc/GMT-9', value: 'Etc/GMT-9' },
    { text: '(~UTC+09:00) Japan', value: 'Japan' },
    { text: '(~UTC+09:00) Pacific/Palau', value: 'Pacific/Palau', openwrt: 'PWT-9' },
    { text: '(~UTC+09:00) ROK', value: 'ROK' },
    { text: '(~UTC+09:00) JST', value: 'JST' },
    { text: '(~UTC+09:30) Australia/Darwin', value: 'Australia/Darwin', openwrt: 'CST-9:30' },
    { text: '(~UTC+09:30) Australia/North', value: 'Australia/North' },
    { text: '(~UTC+09:30) ACT', value: 'ACT' },
    { text: '(~UTC+10:00) Antarctica/DumontDUrville', value: 'Antarctica/DumontDUrville', openwrt: 'DDUT-10' },
    { text: '(~UTC+10:00) Asia/Ust-Nera', value: 'Asia/Ust-Nera' },
    {
      text: '(~UTC+10:00) Asia/Vladivostok',
      value: 'Asia/Vladivostok',
      openwrt: 'VLAT-10VLAST,M3.5.0,M10.5.0/3',
    },
    { text: '(~UTC+10:00) Australia/Brisbane', value: 'Australia/Brisbane', openwrt: 'EST-10' },
    { text: '(~UTC+10:00) Australia/Lindeman', value: 'Australia/Lindeman', openwrt: 'EST-10' },
    { text: '(~UTC+10:00) Australia/Queensland', value: 'Australia/Queensland' },
    { text: '(~UTC+10:00) Etc/GMT-10', value: 'Etc/GMT-10' },
    { text: '(~UTC+10:00) Pacific/Chuuk', value: 'Pacific/Chuuk' },
    { text: '(~UTC+10:00) Pacific/Guam', value: 'Pacific/Guam', openwrt: 'ChST-10' },
    { text: '(~UTC+10:00) Pacific/Port_Moresby', value: 'Pacific/Port_Moresby', openwrt: 'PGT-10' },
    { text: '(~UTC+10:00) Pacific/Saipan', value: 'Pacific/Saipan', openwrt: 'ChST-10' },
    { text: '(~UTC+10:00) Pacific/Truk', value: 'Pacific/Truk', openwrt: 'TRUT-10' },
    { text: '(~UTC+10:00) Pacific/Yap', value: 'Pacific/Yap' },
    {
      text: '(~UTC+10:30) Australia/Adelaide',
      value: 'Australia/Adelaide',
      openwrt: 'CST-9:30CST,M10.1.0,M4.1.0/3',
    },
    {
      text: '(~UTC+10:30) Australia/Broken_Hill',
      value: 'Australia/Broken_Hill',
      openwrt: 'CST-9:30CST,M10.1.0,M4.1.0/3',
    },
    { text: '(~UTC+10:30) Australia/South', value: 'Australia/South' },
    { text: '(~UTC+10:30) Australia/Yancowinna', value: 'Australia/Yancowinna' },
    { text: '(~UTC+11:00) Antarctica/Macquarie', value: 'Antarctica/Macquarie', openwrt: 'MIST-11' },
    { text: '(~UTC+11:00) Asia/Magadan', value: 'Asia/Magadan', openwrt: 'MAGT-11MAGST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+11:00) Asia/Sakhalin', value: 'Asia/Sakhalin', openwrt: 'SAKT-10SAKST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+11:00) Asia/Srednekolymsk', value: 'Asia/Srednekolymsk' },
    { text: '(~UTC+11:00) Australia/ACT', value: 'Australia/ACT' },
    { text: '(~UTC+11:00) Australia/Canberra', value: 'Australia/Canberra' },
    { text: '(~UTC+11:00) Australia/Currie', value: 'Australia/Currie', openwrt: 'EST-10EST,M10.1.0,M4.1.0/3' },
    { text: '(~UTC+11:00) Australia/Hobart', value: 'Australia/Hobart', openwrt: 'EST-10EST,M10.1.0,M4.1.0/3' },
    { text: '(~UTC+11:00) Australia/LHI', value: 'Australia/LHI' },
    {
      text: '(~UTC+11:00) Australia/Lord_Howe',
      value: 'Australia/Lord_Howe',
      openwrt: 'LHST-10:30LHST-11,M10.1.0,M4.1.0',
    },
    {
      text: '(~UTC+11:00) Australia/Melbourne',
      value: 'Australia/Melbourne',
      openwrt: 'EST-10EST,M10.1.0,M4.1.0/3',
    },
    { text: '(~UTC+11:00) Australia/NSW', value: 'Australia/NSW' },
    { text: '(~UTC+11:00) Australia/Sydney', value: 'Australia/Sydney', openwrt: 'EST-10EST,M10.1.0,M4.1.0/3' },
    { text: '(~UTC+11:00) Australia/Tasmania', value: 'Australia/Tasmania' },
    { text: '(~UTC+11:00) Australia/Victoria', value: 'Australia/Victoria' },
    { text: '(~UTC+11:00) Etc/GMT-11', value: 'Etc/GMT-11' },
    { text: '(~UTC+11:00) Pacific/Bougainville', value: 'Pacific/Bougainville' },
    { text: '(~UTC+11:00) Pacific/Efate', value: 'Pacific/Efate', openwrt: 'VUT-11' },
    { text: '(~UTC+11:00) Pacific/Guadalcanal', value: 'Pacific/Guadalcanal', openwrt: 'SBT-11' },
    { text: '(~UTC+11:00) Pacific/Kosrae', value: 'Pacific/Kosrae', openwrt: 'KOST-11' },
    { text: '(~UTC+11:00) Pacific/Norfolk', value: 'Pacific/Norfolk', openwrt: 'NFT-11:30' },
    { text: '(~UTC+11:00) Pacific/Noumea', value: 'Pacific/Noumea', openwrt: 'NCT-11' },
    { text: '(~UTC+11:00) Pacific/Pohnpei', value: 'Pacific/Pohnpei' },
    { text: '(~UTC+11:00) Pacific/Ponape', value: 'Pacific/Ponape', openwrt: 'PONT-11' },
    { text: '(~UTC+11:00) AET', value: 'AET' },
    { text: '(~UTC+11:00) SST', value: 'SST' },
    { text: '(~UTC+12:00) Asia/Anadyr', value: 'Asia/Anadyr', openwrt: 'ANAT-11ANAST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+12:00) Asia/Kamchatka', value: 'Asia/Kamchatka', openwrt: 'PETT-11PETST,M3.5.0,M10.5.0/3' },
    { text: '(~UTC+12:00) Etc/GMT-12', value: 'Etc/GMT-12' },
    { text: '(~UTC+12:00) Kwajalein', value: 'Kwajalein' },
    { text: '(~UTC+12:00) Pacific/Fiji', value: 'Pacific/Fiji', openwrt: 'FJT-12' },
    { text: '(~UTC+12:00) Pacific/Funafuti', value: 'Pacific/Funafuti', openwrt: 'TVT-12' },
    { text: '(~UTC+12:00) Pacific/Kwajalein', value: 'Pacific/Kwajalein', openwrt: 'MHT-12' },
    { text: '(~UTC+12:00) Pacific/Majuro', value: 'Pacific/Majuro', openwrt: 'MHT-12' },
    { text: '(~UTC+12:00) Pacific/Nauru', value: 'Pacific/Nauru', openwrt: 'NRT-12' },
    { text: '(~UTC+12:00) Pacific/Tarawa', value: 'Pacific/Tarawa', openwrt: 'GILT-12' },
    { text: '(~UTC+12:00) Pacific/Wake', value: 'Pacific/Wake', openwrt: 'WAKT-12' },
    { text: '(~UTC+12:00) Pacific/Wallis', value: 'Pacific/Wallis', openwrt: 'WFT-12' },
    {
      text: '(~UTC+13:00) Antarctica/McMurdo',
      value: 'Antarctica/McMurdo',
      openwrt: 'NZST-12NZDT,M9.5.0,M4.1.0/3',
    },
    {
      text: '(~UTC+13:00) Antarctica/South_Pole',
      value: 'Antarctica/South_Pole',
      openwrt: 'NZST-12NZDT,M9.5.0,M4.1.0/3',
    },
    { text: '(~UTC+13:00) Etc/GMT-13', value: 'Etc/GMT-13' },
    { text: '(~UTC+13:00) NZ', value: 'NZ' },
    { text: '(~UTC+13:00) Pacific/Auckland', value: 'Pacific/Auckland', openwrt: 'NZST-12NZDT,M9.5.0,M4.1.0/3' },
    { text: '(~UTC+13:00) Pacific/Enderbury', value: 'Pacific/Enderbury', openwrt: 'PHOT-13' },
    { text: '(~UTC+13:00) Pacific/Fakaofo', value: 'Pacific/Fakaofo', openwrt: 'TKT10' },
    { text: '(~UTC+13:00) Pacific/Tongatapu', value: 'Pacific/Tongatapu', openwrt: 'TOT-13' },
    { text: '(~UTC+13:00) NST', value: 'NST' },
    { text: '(~UTC+13:45) NZ-CHAT', value: 'NZ-CHAT' },
    {
      text: '(~UTC+13:45) Pacific/Chatham',
      value: 'Pacific/Chatham',
      openwrt: 'CHAST-12:45CHADT,M9.5.0/2:45,M4.1.0/3:45',
    },
    { text: '(~UTC+14:00) Etc/GMT-14', value: 'Etc/GMT-14' },
    { text: '(~UTC+14:00) Pacific/Apia', value: 'Pacific/Apia', openwrt: 'WST11' },
    { text: '(~UTC+14:00) Pacific/Kiritimati', value: 'Pacific/Kiritimati', openwrt: 'LINT-14' },
    { text: '(~UTC+14:00) MIT', value: 'MIT' },
  ],
}

export default dates
