--- build-tree/apache2/support/apachectl.in.old	2004-08-30 07:33:35.000000000 -0600
+++ build-tree/apache2/support/apachectl.in	2004-08-30 07:34:30.000000000 -0600
@@ -50,7 +50,7 @@
 # a command that outputs a formatted text version of the HTML at the
 # url given on the command line.  Designed for lynx, however other
 # programs may work.  
-LYNX="lynx -dump"
+LYNX="www-browser -dump"
 #
 # the URL to your server's mod_status status page.  If you do not
 # have one, then status and fullstatus will not work.
