// vuntangle/src/config/theme.js
import { colors } from 'vuetify/lib'
export default {
  theme: {
    dark: false,
    themes: {
      light: {
        primary: '#146095', // arista medium blue
        secondary: '#4473a9', // arista light blue
        accent: '#d47122', // arista orange
        info: '#146095', // arista medium blue
        warning: colors.yellow.darken3,
        error: colors.deepOrange,
        success: colors.green,
        lightGray: '#F1F1F1',
        darkGray: '#5C5C5C',
        utGreen: '#68BD49',
        utDarkBlue: '#303940',
        utBlue: '#0071DC',
        aristaBlue: '#16325B',
        aristaDarkGray: '#58585B',
        aristaLightBlue: '#4473a9',
        aristaLightGray: '#bbbdc0',
        aristaLightGreen: '#aad037',
        aristaMediumBlue: '#146095',
        aristaMediumGray: '#969696',
        aristaOrange: '#d47122',
        aristaYellow: '#e9d554',
        anchor: '#146095', // custom anchor tag color
      },
      dark: {
        primary: '#146095', // arista medium blue
        secondary: '#4473a9', // arista light blue
        accent: '#d47122', // arista orange
        info: '#146095', // arista medium blue as info
        warning: colors.yellow.darken3,
        error: colors.deepOrange,
        success: colors.green,
        lightGray: '#F1F1F1',
        darkGray: '#5C5C5C',
        utGreen: '#68BD49',
        utDarkBlue: '#303940',
        utBlue: '#0071DC',
        aristaBlue: '#16325B',
        aristaDarkGray: '#58585B',
        aristaLightBlue: '#4473a9',
        aristaLightGray: '#bbbdc0',
        aristaLightGreen: '#aad037',
        aristaMediumBlue: '#146095',
        aristaMediumGray: '#969696',
        aristaOrange: '#d47122',
        aristaYellow: '#e9d554',
        anchor: '#146095',
      },
    },
  },
}
