// vuntangle/src/shared/Reports/configs/constants.js
/**
 * Views categories
 */
const Category = Object.freeze({
  Bandwidth: 'bandwidth',
  WanPerformance: 'wan_performance',
  Sessions: 'sessions',
  ThreatPrevention: 'threat_prevention',
  DynamicLists: 'dynamic_blocklist',
  WebFilter: 'web_filter',
  GeoipFilter: 'geoip_filter',
  ApplicationControl: 'application_control',
  CaptivePortal: 'captive_portal',
  PolicyManager: 'policy_manager',
  DenialOfService: 'denial_of_service',
})

/**
 * Id's used to identify a view via route params coming from host app
 */
const View = Object.freeze({
  BandwidthClients: 'bandwidth-clients',
  BandwidthServerPorts: 'bandwidth-server-ports',
  BandwidthApplications: 'bandwidth-applications',
  WanPerformanceUsage: 'wan-performance-usage',
  WanPerformancePacketLoss: 'wan-performance-packet-loss',
  WanPerformanceLatency: 'wan-performance-latency',
  WanPerformanceJitter: 'wan-performance-jitter',
  SessionsList: 'sessions-list',
  SessionsWanRouting: 'sessions-wan-routing',
  ThreatPreventionBlockedSummary: 'threat-prevention-blocked-summary',
  ThreatPreventionBlockedAddresses: 'threat-prevention-blocked-addresses',
  WebFilterSummary: 'web-filter-summary',
  WebFilterVisitedSites: 'web-filter-visited-sites',
  GeoipBlockedSessions: 'geoip-blocked-sessions',
  GeoipBlockedCountriesOutbound: 'geoip-blocked-outbound',
  GeoipBlockedCountriesInbound: 'geoip-blocked-inbound',
  GeoipDistributionMap: 'geoip-distribution-map',
  ApplicationControlApplications: 'application-control-applications',
  CaptivePortalSummary: 'captive-portal-summary',
  CaptivePortalEvents: 'captive-portal-events',
  DynamicListsBlockedSummary: 'dynamic-lists-blocked-summary',
  DynamicListsBlockedAddresses: 'dynamic-lists-blocked-addresses',
  PolicyManagerEvents: 'policy-manager-events',
  DenialOfServiceEvents: 'denial-of-service-events',
})

/**
 * Report ids used in views configurations
 */
const Report = Object.freeze({
  // Bandwidth
  ClientsBandwidthByTime: 'clients_bandwidth_by_time',
  ClientsByBandwidth: 'clients_by_bandwidth',
  ServerPortsBandwidthByTime: 'server_ports_bandwidth_by_time',
  ServerPortsByBandwidth: 'server_ports_by_bandwidth',
  ApplicationsBandwidthByTime: 'applications_bandwidth_by_time',
  ApplicationsByBandwidth: 'applications_by_bandwidth',

  // WAN Performance
  WanPerformanceUsage: 'wan_performance_usage',
  WanPerformancePacketLoss: 'wan_performance_packet_loss',
  WanPerformanceLatency: 'wan_performance_latency',
  WanPerformanceJitter: 'wan_performance_jitter',

  // Sessions
  SessionsList: 'sessions_all',
  SessionsWanRouting: 'sessions_wan_routing',

  // Threat Prevention
  ThreatPreventionBlockedIpsByTime: 'threat_prevention_blocked_ips_by_time',
  ThreatPreventionByBlockedCount: 'threat_prevention_by_blocked_count',
  ThreatPreventionBlockedAddresses: 'threat_prevention_blocked_addresses',

  // Dynamic Lists
  DynamicListsBlockedIpsByTime: 'dynamic_lists_blocked_ips_by_time',
  DynamicListsByBlockedCount: 'dynamic_lists_by_blocked_count',
  DynamicListsBlockedAddresses: 'dynamic_lists_blocked_addresses',

  // Web Filter
  WebFilterTopBlockedSites: 'web_filter_top_blocked_sites',
  WebFilterTopVisitedSites: 'web_filter_top_visited_sites',
  WebFilterTopBlockedCategories: 'web_filter_top_blocked_categories',
  WebFilterTopVisitedCategories: 'web_filter_top_visited_categories',
  WebFilterTopClientsByBlockedSites: 'web_filter_top_clients_by_blocked_sites',
  WebFilterTopClientsByVisitedSites: 'web_filter_top_clients_by_visited_sites',
  WebFilterVisitedSites: 'web_filter_visited_sites',

  // Geoip Filter
  GeoipOutboundBlockedSessions: 'geoip_outbound_blocked_sessions',
  GeoipInboundBlockedSessions: 'geoip_inbound_blocked_sessions',
  GeoipOutboundBlockedCountriesByTime: 'geoip_outbound_blocked_countries_by_time',
  GeoipOutboundCountriesByBlockCount: 'geoip_outbound_countries_by_blocked_count',
  GeoipInboundBlockedCountriesByTime: 'geoip_inbound_blocked_countries_by_time',
  GeoipInboundCountriesByBlockCount: 'geoip_inbound_countries_by_blocked_count',
  GeoipDistributionMap: 'geoip_distribution_map',

  // Application Control
  ApplicationControlApplications: 'application_control_applications',

  // Captive Portal
  CaptivePortalByTime: 'captive_portal_by_time',
  CaptivePortalByCount: 'captive_portal_by_count',
  CaptivePortalEvents: 'captive_portal_events',

  // Policy Manager
  PolicyManagerEvents: 'policy_manager_events',

  // Denial of service
  DenialOfServiceEvents: 'denial_of_service_events',
})

export { Category, View, Report }
