diff -urN debian.old/build/iptables-1.2.9/ip6tables.c debian/build/iptables-1.2.8/ip6tables.c
--- debian.old/build/iptables-1.2.9/ip6tables.c	2003-03-05 02:46:15.000000000 -0500
+++ debian/build/iptables-1.2.9/ip6tables.c	2003-06-21 07:42:26.000000000 -0400
@@ -841,7 +841,7 @@
 			if (!isalnum(vianame[i]) 
 			    && vianame[i] != '_' 
 			    && vianame[i] != '.') {
-				printf("Warning: wierd character in interface"
+				printf("Warning: weird character in interface"
 				       " `%s' (No aliases, :, ! or *).\n",
 				       vianame);
 				break;
diff -urN debian.old/build/iptables-1.2.9/iptables.c debian/build/iptables-1.2.8/iptables.c
--- debian.old/build/iptables-1.2.9/iptables.c	2003-03-31 07:11:55.000000000 -0500
+++ debian/build/iptables-1.2.9/iptables.c	2003-06-21 07:42:26.000000000 -0400
@@ -782,7 +782,7 @@
 			if (!isalnum(vianame[i]) 
 			    && vianame[i] != '_' 
 			    && vianame[i] != '.') {
-				printf("Warning: wierd character in interface"
+				printf("Warning: weird character in interface"
 				       " `%s' (No aliases, :, ! or *).\n",
 				       vianame);
 				break;
