// vuntangle/src/policy-manager/config/objectsConfig.js
/**
 * Object configurations based on Type
 * [Type.Name]: {
 *   objectType: optional, used if it is different than the Type.Name (like uno/agni)
 *   linkedObjectType: additional object type to load from the store, typically used when adding an
 *                     object to an object group
 *   editComponent: `ComponentName` => the component used for editing object
 *   text: 'some_text' => the translaiton key used for it
 *   defaults: ['some_objects'] => the array that stores the default configurations for the objects
 *   autoHideGeneratedObjects: true => hides any object starting with name "Autogenerated "
 * }
 */

import { Type } from './constants'

export default Object.freeze({
  [Type.ObjectCondition]: {
    text: 'conditions',
    autoHideGeneratedObjects: true,
  },
  [Type.ObjectConditionGroup]: {
    linkedObjectType: Type.ObjectCondition,
    editComponent: 'GroupsEdit',
    text: 'condition_group',
  },

  [Type.ObjectApplicationControlList]: {
    editComponent: 'ObjectApplication',
    text: 'applications_control_list',
    hidden: true,
  },
  [Type.ObjectApplicationControlListGroup]: {
    linkedObjectType: Type.ObjectApplicationControlList,
    editComponent: 'GroupsEdit',
    text: 'application_control_list_group',
    hidden: true,
  },

  [Type.ObjectApplication]: {
    editComponent: 'ObjectApplication',
    text: 'application',
    readOnly: false,
  },
  [Type.ObjectApplicationGroup]: {
    linkedObjectType: Type.ObjectApplication,
    editComponent: 'GroupsEdit',
    text: 'application_groups',
    readOnly: false,
  },

  [Type.ObjectGeoip]: {
    editComponent: 'ObjectGeoip',
    text: 'geoip_location',
  },

  [Type.ObjectGeoipGroup]: {
    linkedObjectType: Type.ObjectGeoip,
    editComponent: 'GroupsEdit',
    text: 'geoip_location_group',
  },

  [Type.ObjectHostname]: {
    editComponent: 'ObjectHostname',
    text: 'hostname',
  },

  [Type.ObjectHostnameGroup]: {
    linkedObjectType: Type.ObjectHostname,
    editComponent: 'GroupsEdit',
    text: 'hostname_groups',
  },

  [Type.ObjectInterfaceZone]: {
    editComponent: 'ObjectInterfaceZone',
    text: 'interface_zone',
  },

  [Type.ObjectInterfaceZoneGroup]: {
    linkedObjectType: Type.ObjectInterfaceZone,
    editComponent: 'GroupsEdit',
    text: 'interface_zone_group',
  },

  [Type.ObjectIpAddress]: {
    editComponent: 'ObjectIpAddress',
    text: 'ip_address_list',
    showIpCustomSorting: true,
  },

  [Type.ObjectIpAddressGroup]: {
    linkedObjectType: Type.ObjectIpAddress,
    editComponent: 'GroupsEdit',
    text: 'ip_address_list_group',
  },

  [Type.ObjectService]: {
    editComponent: 'ObjectService',
    text: 'service_endpoint',
    defaults: [
      {
        Name: 'Default HTTP',
        Description: 'Default HTTP',
        Type: Type.ObjectService,
        PolicyJson: {
          'items': [
            {
              'port': ['80'],
              'protocol': ['6'],
            },
          ],
          'type': Type.ObjectService,
        },
      },
      {
        Name: 'Default HTTPS',
        Description: 'Default HTTPS',
        Type: Type.ObjectService,
        PolicyJson: {
          'items': [
            {
              'port': ['443'],
              'protocol': ['6'],
            },
          ],
          'type': Type.ObjectService,
        },
      },
      {
        Name: 'Default SSH',
        Description: 'Default SSH',
        Type: Type.ObjectService,
        PolicyJson: {
          'items': [
            {
              'port': ['23'],
              'protocol': ['6'],
            },
          ],
          'type': Type.ObjectService,
        },
      },
      {
        Name: 'Default FTP',
        Description: 'Default FTP',
        Type: Type.ObjectService,
        PolicyJson: {
          'items': [
            {
              'port': ['21'],
              'protocol': ['6'],
            },
          ],
          'type': Type.ObjectService,
        },
      },
      {
        Name: 'Default RDP',
        Description: 'Default RDP',
        Type: Type.ObjectService,
        PolicyJson: {
          'items': [
            {
              'port': ['3389'],
              'protocol': ['6'],
            },
          ],
          'type': Type.ObjectService,
        },
      },
    ],
  },

  [Type.ObjectServiceGroup]: {
    linkedObjectType: Type.ObjectService,
    editComponent: 'GroupsEdit',
    text: 'service_endpoint_group',
  },

  [Type.ObjectUser]: {
    editComponent: 'ObjectUser',
    text: 'user',
    readOnly: false,
  },

  [Type.ObjectUserGroup]: {
    linkedObjectType: Type.ObjectUser,
    editComponent: 'GroupsEdit',
    text: 'user_group',
    readOnly: false,
  },

  [Type.ObjectVlanTag]: {
    editComponent: 'ObjectVlanTag',
    text: 'vlan_tag',
  },
  [Type.ObjectVlanTagGroup]: {
    linkedObjectType: Type.ObjectVlanTag,
    editComponent: 'GroupsEdit',
    text: 'vlan_tag_group',
  },

  [Type.ObjectAgni]: {
    objectType: Type.ObjectUser,
    text: 'agni_user',
    readOnly: true,
    allowDelete: true,
  },

  [Type.ObjectAgniGroup]: {
    objectType: Type.ObjectUserGroup,
    linkedObjectType: Type.ObjectUser,
    text: 'agni_user_groups',
    readOnly: true,
    allowDelete: true,
  },

  [Type.ObjectUno]: {
    objectType: Type.ObjectApplication,
    editComponent: 'ObjectUno',
    text: 'uno_application',
    readOnly: true,
    allowDelete: true,
  },

  [Type.ObjectUnoGroup]: {
    objectType: Type.ObjectApplicationGroup,
    linkedObjectType: Type.ObjectApplication,
    text: 'uno_application_groups',
    readOnly: true,
    allowDelete: true,
  },
})
