// vuntangle/src/policy-manager/config/templatesConfig.js
import { Type } from './constants'

/**
 * Templates configurations based on Type
 * This are used only with Policy Manager templates configurations
 * [Type.Name]: {
 *   text: 'some_text' => the translaiton key used for it
 *   settingsComponent: 'Settings...' => vuntangle component used to edit the template
 *   tooltipComponent: used to render templates in a nice tooltip when editing a rule
 *   defaultSettings: {} => default template settings upon creation
 *   category: 'global' => to which category this config belongs (to build up categorised menu)
 * }
 */

export default Object.freeze({
  [Type.ConfigApplicationControl]: {
    text: 'application_control',
    category: 'policy',
    settingsComponent: 'SettingsApplicationControl',
    tooltipComponent: 'TooltipApplicationControl',
    defaultSettings: {
      enabled: true,
      cloud_classification: true,
      custom_rules: [],
      actions: {
        block: [],
        flag: [],
        reject: [],
      },
    },
  },
  [Type.ConfigCaptivePortal]: {
    text: 'captive_portal',
    category: 'policy',
    settingsComponent: 'SettingsCaptivePortal',
    tooltipComponent: 'TooltipCaptivePortal',
    defaultSettings: {
      enabled: false,
      acceptText: 'I agree',
      acceptButtonText: 'Continue',
      messageHeading: 'Guest Access Notice',
      messageText: 'Click Continue to access the Internet',
      pageTitle: 'Guest access',
      tosText:
        'If you are not eligible, or do not agree to the terms, then you do not have our permission to use the Site. Arista, at its sole discretion, reserves the right to change these terms and conditions at any time. Please check these terms periodically for changes. By using the Site after changes are published, you agree to the updated terms. Use of our products and services requires a separate agreement with us, even if portions may be accessed through this Site.',
      welcomeText: 'Welcome to guest access',
      logo: {
        imageName: '',
        imageData: '',
      },
      timeoutValue: 1,
      timeoutPeriod: 'd',
    },
  },
  [Type.ConfigGeoipFilter]: {
    text: 'geoip_filter',
    category: 'policy',
    settingsComponent: 'SettingsGeoipFilter',
    tooltipComponent: 'TooltipGeoipFilter',
    defaultSettings: {
      blockedCountries: [],
      enabled: true,
      enabledLog: false,
      passedNetworks: [],
    },
  },
  [Type.ConfigThreatPrevention]: {
    text: 'threat_prevention',
    category: 'policy',
    settingsComponent: 'SettingsThreatPrevention',
    tooltipComponent: 'TooltipThreatPrevention',
    defaultSettings: {
      enabled: true,
      passList: [],
      redirect: false,
      sensitivity: 20,
    },
  },
  // CD- 6048 Hide WAN Policies from Policy Manager templates (not yet implemented)
  // [Type.ConfigWanPolicy]: {
  //   text: 'wan_policy',
  //   category: 'policy',
  //   settingsComponent: 'SettingsWanPolicies',
  //   tooltipComponent: 'TooltipWanPolicy',
  //   defaultSettings: {
  //     criteria: [],
  //     interfaces: [{ interfaceId: 0 }],
  //     type: 'BEST_OF',
  //     best_of_metric: 'LOWEST_LATENCY',
  //   },
  // },
  [Type.ConfigWebFilter]: {
    text: 'web_filter',
    category: 'policy',
    settingsComponent: 'SettingsWebFilter',
    tooltipComponent: 'TooltipWebFilter',
    defaultSettings: {
      blockList: [],
      categories: [
        {
          enabled: true,
          flagged: true,
          id: 11,
        },
        {
          enabled: true,
          flagged: true,
          id: 44,
        },
        {
          enabled: true,
          flagged: true,
          id: 49,
        },
        {
          enabled: true,
          flagged: true,
          id: 56,
        },
        {
          enabled: true,
          flagged: true,
          id: 57,
        },
        {
          enabled: true,
          flagged: true,
          id: 58,
        },
        {
          enabled: true,
          flagged: true,
          id: 59,
        },
        {
          enabled: true,
          flagged: true,
          id: 67,
        },
      ],
      enabled: true,
      passList: [],
    },
  },
  [Type.ConfigGlobalDns]: {
    text: 'dns',
    category: 'global',
    settingsComponent: 'SettingsDns',
  },
  [Type.ConfigGlobalStatusAnalyzers]: {
    text: 'status_analyzers',
    category: 'global',
    settingsComponent: 'AnalyzersList',
    defaultSettings: [
      {
        enabled: true,
        interfaceIds: [],
        ipv4Addresses: ['1.1.1.1', '1.0.0.1', '8.8.8.8', '8.8.4.4', '208.67.222.222', '208.67.220.220'],
        ipv6Addresses: [
          '2606:4700:4700::1111',
          '2606:4700:4700::1001',
          '2001:4860:4860::8888',
          '2001:4860:4860::8844',
          '2620:119:35::35',
          '2620:119:53::53',
        ],
        name: 'Default Ping Analyzer',
      },
    ],
  },
  [Type.ConfigGlobalUpgrade]: {
    text: 'upgrade',
    category: 'global',
    settingsComponent: 'ApplianceUpgrade',
    defaultSettings: {
      enabled: true,
      dayOfWeek: 0,
      hourOfDay: 0,
      minuteOfHour: 0,
    },
  },
  [Type.ConfigGlobalDynamicBlocklist]: {
    text: 'dynamic_blocklist',
    category: 'global',
    settingsComponent: 'SettingsDynamicBlockLists',
    defaultSettings: {
      'configurations': [
        {
          'enabled': false,
          'id': 'defa17b1-b10c-4de5-7897-3f426fb32b99',
          'name': 'Emerging Threats',
          'parsingMethod': '^\\S{2,256}',
          'pollingTime': 30,
          'pollingUnit': 'Minutes',
          'skipCertCheck': false,
          'source': 'http://opendbl.net/lists/etknown.list',
          'type': 'IPList',
        },
        {
          'enabled': false,
          'id': 'f9209659-3968-4de5-9385-3f426df3411a',
          'name': 'DShield Blocklist',
          'parsingMethod':
            '((?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)+|(?:[a-f0-9:]+:+)+(?:[a-f0-9](?:(::)?))+)(?:\\/{1}\\d+|-((?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)+|(?:[a-f0-9:]+:+)+(?:[a-f0-9](?:(::)?))+))?',
          'pollingTime': 1,
          'pollingUnit': 'Hours',
          'skipCertCheck': false,
          'source': 'http://opendbl.net/lists/dshield.list',
          'type': 'IPList',
        },
      ],
      enabled: true,
    },
  },
  [Type.ConfigGlobalStaticRoutes]: {
    text: 'static_routes',
    category: 'global',
    settingsComponent: 'StaticRoutes',
    defaultSettings: [],
  },
  [Type.ConfigGlobalNetworkDiscovery]: {
    text: 'network_discovery',
    category: 'global',
    settingsComponent: 'SettingsNetworkDiscovery',
  },
  // [Type.ConfigGlobalDenialOfService]: {
  //   text: 'denial_of_service',
  //   category: 'global',
  //   settingsComponent: 'SettingsDenialOfService',
  // },
  [Type.ConfigGlobalDatabase]: {
    text: 'database',
    category: 'global',
    settingsComponent: 'DatabaseList',
    defaultSettings: [],
  },
})
