// vuntangle/src/shared/SettingsCaptivePortal/conditionDefs.js
import { portProtocolOptionsAsString, isOperatorOptions, protocolOptions } from '../../constants'

export const conditionDefs = {
  SOURCE_ADDRESS: { ops: isOperatorOptions, rules: 'required|ip_expression' },
  SOURCE_PORT: { ops: isOperatorOptions, autocompleteItems: portProtocolOptionsAsString, rules: 'required|port' },
  SOURCE_ADDRESS_V6: { ops: isOperatorOptions, rules: 'required|ip_v6_expression' },
  SOURCE_INTERFACE_ZONE: { field: 'select', ops: isOperatorOptions, rules: 'required' },
  DESTINATION_ADDRESS: { ops: isOperatorOptions, rules: 'required|ip_expression' },
  DESTINATION_PORT: { ops: isOperatorOptions, autocompleteItems: portProtocolOptionsAsString, rules: 'required|port' },
  DESTINATION_ADDRESS_V6: { ops: isOperatorOptions, rules: 'required|ip_v6_expression' },
  IP_PROTOCOL: { ops: isOperatorOptions, autocompleteItems: protocolOptions, rules: 'required' },
}

export const conditionTypeHeaders = [
  { prefix: 'SOURCE', text: 'SOURCE' },
  { prefix: 'DESTINATION', text: 'DESTINATION' },
]

export const conditionTypeOptions = Object.keys(conditionDefs)
