diff -urN --ignore-all-space old/iptables_profectio/extensions/Makefile new/iptables_profectio/extensions/Makefile
--- old/iptables_profectio/extensions/Makefile	2004-06-17 03:22:54.000000000 -0700
+++ new/iptables_profectio/extensions/Makefile	2006-07-17 15:58:00.000000000 -0700
@@ -5,7 +5,8 @@
 # header files are present in the include/linux directory of this iptables
 # package (HW)
 #
-PF_EXT_SLIB:=ah addrtype comment connlimit connmark conntrack dscp ecn esp hashlimit helper icmp iprange length limit mac mark multiport owner physdev pkttype realm rpc sctp standard state tcp tcpmss tos ttl udp unclean CLASSIFY CONNMARK DNAT DSCP ECN LOG MARK MASQUERADE MIRROR NETMAP NFQUEUE NOTRACK REDIRECT REJECT SAME SNAT TARPIT TCPMSS TOS TRACE TTL ULOG
+PF_EXT_SLIB:=ah addrtype comment connlimit connmark conntrack dscp ecn esp hashlimit helper icmp iprange length limit mac mark multiport owner physdev pkttype realm rpc sctp standard state tcp tcpmss tos ttl udp unclean CLASSIFY CONNMARK DIVERT DNAT DSCP ECN LOG MARK MASQUERADE MIRROR NETMAP NFQUEUE NOTRACK REDIRECT REJECT SAME SNAT TARPIT TCPMSS TOS TRACE TTL ULOG
+
 PF6_EXT_SLIB:=eui64 hl icmpv6 length limit mac mark multiport owner physdev standard tcp udp HL LOG NFQUEUE MARK TRACE
 
 # Optionals
--- old/iptables_profectio/extensions/libipt_DIVERT.c	1969-12-31 16:00:00.000000000 -0800
+++ new/iptables_profectio/extensions/libipt_DIVERT.c	2006-07-17 15:58:00.000000000 -0700
@@ -0,0 +1,103 @@
+/* Shared library add-on to iptables to add TPROXY target support. */
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <getopt.h>
+
+#include <iptables.h>
+#include <linux/netfilter_ipv4/ip_tables.h>
+#include <linux/netfilter_ipv4/ipt_DIVERT.h>
+
+struct tosinfo {
+    struct ipt_entry_target t;
+    struct ipt_divert_target_info divert;
+};
+
+static void help(void)
+{
+    printf("DIVERT target v%s options:\n"
+           "  --to-port port                   DIVERT packets to port\n",
+           IPTABLES_VERSION);
+}
+
+static struct option opts[] = {
+    { "to-port", 1, 0, '1' },
+    { 0 }
+};
+
+static void init(struct ipt_entry_target *t, unsigned int *nfcache)
+{}
+
+static void parse_divert(const unsigned char *s, struct ipt_divert_target_info *info)
+{
+    int divert_port;
+    string_to_number(s, 0, 65535, &divert_port);
+    info->to_port = htons(divert_port);
+}
+
+static int parse(int c, char **argv, int invert, unsigned int *flags,
+                 const struct ipt_entry *entry,
+                 struct ipt_entry_target **target)
+{
+    struct ipt_divert_target_info *divertinfo
+        = (struct ipt_divert_target_info *)(*target)->data;
+
+    switch (c) {
+    case '1':
+        if (*flags)
+            exit_error(PARAMETER_PROBLEM,"DIVERT target: Can't specify --to-port twice");
+        parse_divert(optarg, divertinfo);
+        *flags = 1;
+        break;
+
+    default:
+        return 0;
+    }
+
+    return 1;
+}
+
+static void final_check(unsigned int flags)
+{
+    if (!flags)
+        exit_error(PARAMETER_PROBLEM,
+                   "DIVERT target: Parameter --to-port is required");
+}
+
+static void print(const struct ipt_ip *ip,
+                  const struct ipt_entry_target *target,
+                  int numeric)
+{
+    const struct ipt_divert_target_info *divertinfo =
+        (const struct ipt_divert_target_info *)target->data;
+    printf("DIVERT redirect %d", ntohs(divertinfo->to_port));
+}
+
+static void save(const struct ipt_ip *ip, const struct ipt_entry_target *target)
+{
+    const struct ipt_divert_target_info *divertinfo =
+        (const struct ipt_divert_target_info *)target->data;
+
+    printf("--to-port %d ", ntohs(divertinfo->to_port));
+}
+
+static 
+struct iptables_target divert 
+= { .next           = NULL,
+    .name           = "DIVERT",
+    .version        = IPTABLES_VERSION,
+    .size           = IPT_ALIGN(sizeof(struct ipt_divert_target_info)),
+    .userspacesize  = IPT_ALIGN(sizeof(struct ipt_divert_target_info)),
+    .help           = &help,
+    .init           = &init,
+    .parse          = &parse,
+    .final_check    = &final_check,
+    .print          = &print,
+    .save           = &save,
+    .extra_opts     = opts
+};
+
+void _init(void)
+{
+    register_target(&divert);
+}
diff -urN --ignore-all-space old/iptables_profectio/include/linux/netfilter_ipv4/ipt_DIVERT.h new/iptables_profectio/include/linux/netfilter_ipv4/ipt_DIVERT.h
--- old/iptables_profectio/include/linux/netfilter_ipv4/ipt_DIVERT.h	1969-12-31 16:00:00.000000000 -0800
+++ new/iptables_profectio/include/linux/netfilter_ipv4/ipt_DIVERT.h	2006-07-17 15:58:00.000000000 -0700
@@ -0,0 +1,9 @@
+/* $Id: netcap.iptables-1.2.9.091604.patch,v 1.1.1.1 2005/01/11 22:59:32 dmorris Exp $ */
+#ifndef _IPT_DIVERT_H_target
+#define _IPT_DIVERT_H_target
+
+struct ipt_divert_target_info {
+	unsigned short to_port;
+};
+
+#endif
diff -Naur old/iptables_profectio/extensions/.policy-test new/iptables_profectio/extensions/.policy-test
--- old/iptables_profectio/extensions/.policy-test  2006-10-05 18:40:38.000000000 -0700
+++ new/iptables_profectio/extensions/.policy-test  1969-12-31 16:00:00.000000000 -0800
@@ -1,3 +0,0 @@
-#!/bin/sh
-#
-[ -f $KERNEL_DIR/include/linux/netfilter_ipv4/ipt_policy.h ] && echo policy
diff -Naur old/iptables_profectio/extensions/.policy-test6 new/iptables_profectio/extensions/.policy-test6
--- old/iptables_profectio/extensions/.policy-test6 2006-10-05 18:40:38.000000000 -0700
+++ new/iptables_profectio/extensions/.policy-test6 1969-12-31 16:00:00.000000000 -0800
@@ -1,3 +0,0 @@
-#!/bin/sh
-#
-[ -f $KERNEL_DIR/include/linux/netfilter_ipv6/ip6t_policy.h ] && echo policy
diff -Naur old/iptables_profectio/extensions/.string-test new/iptables_profectio/.string-test
--- old/iptables_profectio/extensions/.string-test  2004-10-10 02:56:26.000000000 -0700
+++ new/iptables_profectio/extensions/.string-test  1969-12-31 16:00:00.000000000 -0800
@@ -1,2 +0,0 @@
-#! /bin/sh
-[ -f $KERNEL_DIR/include/linux/netfilter_ipv4/ipt_string.h ] && echo string
