// vuntangle/src/shared/SettingsWebFilter/schema.js
export default {
  title: 'Web Filter schema',
  type: 'object',
  properties: {
    enabled: { type: 'boolean' },
    blockList: {
      type: 'array',
      items: {
        type: 'object',
        properties: {
          name: { type: 'string' }, // site name
          description: { type: 'string' },
          enabled: { type: 'boolean' },
          exact: { type: 'boolean' },
          flagged: { type: 'boolean' },
        },
        required: ['name', 'enabled', 'exact', 'flagged'],
        additionalProperties: false,
      },
    },
    passList: {
      type: 'array',
      items: {
        type: 'object',
        properties: {
          name: { type: 'string' }, // site name
          description: { type: 'string' },
          enabled: { type: 'boolean' },
          exact: { type: 'boolean' },
          flagged: { type: 'boolean' },
        },
        required: ['name', 'enabled', 'exact', 'flagged'],
        additionalProperties: false,
      },
    },
    categories: {
      type: 'array',
      items: {
        type: 'object',
        properties: {
          enabled: { type: 'boolean' },
          flagged: { type: 'boolean' },
          id: { type: 'number' }, // category id, see categories.json
        },
        required: ['enabled', 'flagged', 'id'],
        additionalProperties: false,
      },
    },
  },
  required: ['enabled', 'blockList', 'passList', 'categories'],
  additionalProperties: false,
}
