// vuntangle/src/shared/SettingsApplicationControl/schema.js
export default {
  title: 'Application control',
  type: 'object',
  properties: {
    enabled: {
      type: 'boolean',
    },
    cloud_classification: {
      type: 'boolean',
    },
    custom_rules: {
      type: 'array',
      items: {
        type: 'object',
        properties: {
          name: {
            type: 'string',
          },
          description: {
            type: 'string',
          },
          category: {
            type: 'string',
          },
          conditions: {
            type: 'array',
            items: {
              type: 'object',
              properties: {
                type: {
                  type: 'string',
                  enum: ['CLIENT_ADDRESS', 'CLIENT_PORT', 'SERVER_ADDRESS', 'SERVER_PORT', 'PROTOCOL_TYPE'],
                },
                value: {
                  type: 'string',
                },
                op: {
                  type: 'string',
                  enum: ['==', '!=', '<', '>', '<=', '>='],
                },
              },
              required: ['type', 'value', 'op'],
            },
          },
          action: {
            type: 'string',
          },
        },
        required: ['name', 'description', 'category', 'conditions', 'action'],
      },
    },
    actions: {
      type: 'object',
      properties: {
        reject: {
          type: 'array',
          items: {
            type: 'string',
          },
        },
        block: {
          type: 'array',
          items: {
            type: 'string',
          },
        },
        flag: {
          type: 'array',
          items: {
            type: 'string',
          },
        },
      },
      required: ['reject', 'block', 'flag'],
    },
  },
  required: ['enabled', 'cloud_classification', 'custom_rules'],
  additionalProperties: false,
}
