// vuntangle/src/shared/index.js
import { priorities, priorityOptions } from '../constants'
import SettingsApplicationControl from './SettingsApplicationControl/Main.vue'
import SettingsGeoipFilter from './SettingsGeoipFilter/Main.vue'
import SettingsNetworkDiscovery from './SettingsNetworkDiscovery/Main.vue'
import SettingsThreatPrevention from './SettingsThreatPrevention/Main.vue'
import SettingsCaptivePortal from './SettingsCaptivePortal/Main.vue'
import SettingsWebFilter from './SettingsWebFilter/Main.vue'
import webFilterCategories from './SettingsWebFilter/categories.json'

import RulesList from './Rules/RulesList.vue'
import Rule from './Rules/Rule.vue'
import NoLicense from './NoLicense/Main.vue'
import WanPoliciesList from './WanPolicies/WanPoliciesList.vue'
import WanPolicy from './WanPolicies/WanPolicy.vue'
import AnalyzersList from './StatusAnalyzers/AnalyzersList.vue'
import SettingsInterface from './SettingsInterface/Main.vue'

import Dhcp from './SettingsDhcp/Main.vue'
import Dns from './SettingsDns/Main'
import Interfaces from './Interfaces/Interfaces.vue'
import SettingsDynamicBlockLists from './SettingsDynamicBlockLists/DynamicBlockLists.vue'
import ActiveSessions from './ActiveSessions/ActiveSessions.vue'
import Reports from './Reports/Reports.vue'
import ReportDetails from './Reports/ReportDetails.vue'
import ApplianceClients from './Clients/Main.vue'
import unsavedChangesMixin from './unsavedChangesMixin'

// policy manager
import DnsSchema from './SettingsDns/schema'
import ShapingSchema from './SettingsShaping/schema'
import FilterSchema from './SettingsFilter/schema'
import ApplicationControlSchema from './SettingsApplicationControl/schema'
import CaptivePortalSchema from './SettingsCaptivePortal/schema'
import GeoipSchema from './SettingsGeoipFilter/schema'
import NetworkDiscoverySchema from './SettingsNetworkDiscovery/schema'
import ThreatPreventionSchema from './SettingsThreatPrevention/schema'
import WebFilterSchema from './SettingsWebFilter/schema'
import DhcpSchema from './SettingsDhcp/schema'
import WanPolicySchema from './SettingsWanPolicies/schema'
import WanRulesSchema from './SettingsWanRules/schema'
import AccessSchema from './SettingsAccess/schema'
import ApplianceLogging from './ApplianceLogging/Main.vue'
import ApplianceSystem from './ApplianceSystem/Main.vue'
import ApplianceUpgrade from './ApplianceUpgrade/Main.vue'
import TimeSelector from './ApplianceUpgrade/TimeSelector.vue'

import StaticRoutes from './StaticRoutes/StaticRoutes.vue'
import SettingsDenialOfService from './SettingsDenialOfService/Main.vue'
import DatabaseList from './SettingsDatabase/DatabaseList.vue'

import { filterDefaultRules, accessDefaultRules, wanDefaultRules, shapingDefaultRules } from './Rules/defaults'

const SettingSchemas = {
  'dns': DnsSchema,
  'shaping': ShapingSchema,
  'wan-policies': WanPolicySchema,
  'wan-rules': WanRulesSchema,
  'filter': FilterSchema,
  'access': AccessSchema,
  'application-control': ApplicationControlSchema,
  'captive-portal': CaptivePortalSchema,
  'geoip-filter': GeoipSchema,
  'network-discovery': NetworkDiscoverySchema,
  'threat-prevention': ThreatPreventionSchema,
  'web-filter': WebFilterSchema,
  'dhcp': DhcpSchema,
}
export {
  ActiveSessions,
  Reports,
  ReportDetails,
  SettingsApplicationControl,
  SettingsGeoipFilter,
  SettingsNetworkDiscovery,
  SettingsThreatPrevention,
  SettingsCaptivePortal,
  SettingsWebFilter,
  webFilterCategories,
  SettingsInterface,
  RulesList,
  Rule,
  WanPoliciesList,
  WanPolicy,
  WanPolicy as SettingsWanPolicies,
  AnalyzersList,
  Dhcp,
  Dhcp as SettingsDhcp,
  Dns,
  Dns as SettingsDns,
  SettingsDynamicBlockLists,
  NoLicense,
  filterDefaultRules,
  accessDefaultRules,
  wanDefaultRules,
  shapingDefaultRules,
  SettingSchemas,
  Interfaces,
  ApplianceLogging,
  ApplianceSystem,
  ApplianceClients,
  unsavedChangesMixin,
  ApplianceUpgrade,
  TimeSelector,
  StaticRoutes,
  DatabaseList,
  SettingsDenialOfService,
  priorities,
  priorityOptions,
}
